<!DOCTYPE html>
<html lang="en"><head>
<script src="index_files/libs/clipboard/clipboard.min.js"></script>
<script src="index_files/libs/quarto-html/tabby.min.js"></script>
<script src="index_files/libs/quarto-html/popper.min.js"></script>
<script src="index_files/libs/quarto-html/tippy.umd.min.js"></script>
<link href="index_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="index_files/libs/quarto-html/light-border.css" rel="stylesheet">
<link href="index_files/libs/quarto-html/quarto-html.min.css" rel="stylesheet" data-mode="light">
<link href="index_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="index_files/libs/quarto-contrib/reveal-header-1.0.0/add_header.js" defer="true"></script>
<link href="index_files/libs/quarto-contrib/reveal-header-1.0.0/add_header.css" rel="stylesheet">
<script src="index_files/libs/quarto-contrib/sc-sb-title-1.0.0/sc_sb_title.js" defer="true"></script>
<link href="index_files/libs/quarto-contrib/grid-no-htext-1.0.0/grid_no_htext.css" rel="stylesheet"><meta charset="utf-8">
  <meta name="generator" content="quarto-1.4.549">

  <title>index</title>
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no, minimal-ui">
  <link rel="stylesheet" href="index_files/libs/revealjs/dist/reset.css">
  <link rel="stylesheet" href="index_files/libs/revealjs/dist/reveal.css">
  <style>
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    div.columns{display: flex; gap: min(4vw, 1.5em);}
    div.column{flex: auto; overflow-x: auto;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    ul.task-list{list-style: none;}
    ul.task-list li input[type="checkbox"] {
      width: 0.8em;
      margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
      vertical-align: middle;
    }
    /* CSS for syntax highlighting */
    pre > code.sourceCode { white-space: pre; position: relative; }
    pre > code.sourceCode > span { line-height: 1.25; }
    pre > code.sourceCode > span:empty { height: 1.2em; }
    .sourceCode { overflow: visible; }
    code.sourceCode > span { color: inherit; text-decoration: inherit; }
    div.sourceCode { margin: 1em 0; }
    pre.sourceCode { margin: 0; }
    @media screen {
    div.sourceCode { overflow: auto; }
    }
    @media print {
    pre > code.sourceCode { white-space: pre-wrap; }
    pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
    }
    pre.numberSource code
      { counter-reset: source-line 0; }
    pre.numberSource code > span
      { position: relative; left: -4em; counter-increment: source-line; }
    pre.numberSource code > span > a:first-child::before
      { content: counter(source-line);
        position: relative; left: -1em; text-align: right; vertical-align: baseline;
        border: none; display: inline-block;
        -webkit-touch-callout: none; -webkit-user-select: none;
        -khtml-user-select: none; -moz-user-select: none;
        -ms-user-select: none; user-select: none;
        padding: 0 4px; width: 4em;
      }
    pre.numberSource { margin-left: 3em;  padding-left: 4px; }
    div.sourceCode
      { color: #24292e;  }
    @media screen {
    pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
    }
    code span { color: #24292e; } /* Normal */
    code span.al { color: #ff5555; font-weight: bold; } /* Alert */
    code span.an { color: #6a737d; } /* Annotation */
    code span.at { color: #d73a49; } /* Attribute */
    code span.bn { color: #005cc5; } /* BaseN */
    code span.bu { color: #d73a49; } /* BuiltIn */
    code span.cf { color: #d73a49; } /* ControlFlow */
    code span.ch { color: #032f62; } /* Char */
    code span.cn { color: #005cc5; } /* Constant */
    code span.co { color: #6a737d; } /* Comment */
    code span.cv { color: #6a737d; } /* CommentVar */
    code span.do { color: #6a737d; } /* Documentation */
    code span.dt { color: #d73a49; } /* DataType */
    code span.dv { color: #005cc5; } /* DecVal */
    code span.er { color: #ff5555; text-decoration: underline; } /* Error */
    code span.ex { color: #d73a49; font-weight: bold; } /* Extension */
    code span.fl { color: #005cc5; } /* Float */
    code span.fu { color: #6f42c1; } /* Function */
    code span.im { color: #032f62; } /* Import */
    code span.in { color: #6a737d; } /* Information */
    code span.kw { color: #d73a49; } /* Keyword */
    code span.op { color: #24292e; } /* Operator */
    code span.ot { color: #6f42c1; } /* Other */
    code span.pp { color: #d73a49; } /* Preprocessor */
    code span.re { color: #6a737d; } /* RegionMarker */
    code span.sc { color: #005cc5; } /* SpecialChar */
    code span.ss { color: #032f62; } /* SpecialString */
    code span.st { color: #032f62; } /* String */
    code span.va { color: #e36209; } /* Variable */
    code span.vs { color: #032f62; } /* VerbatimString */
    code span.wa { color: #ff5555; } /* Warning */
    /* CSS for citations */
    div.csl-bib-body { }
    div.csl-entry {
      clear: both;
    }
    .hanging-indent div.csl-entry {
      margin-left:2em;
      text-indent:-2em;
    }
    div.csl-left-margin {
      min-width:2em;
      float:left;
    }
    div.csl-right-inline {
      margin-left:2em;
      padding-left:1em;
    }
    div.csl-indent {
      margin-left: 2em;
    }  </style>
  <link rel="stylesheet" href="index_files/libs/revealjs/dist/theme/quarto.css">
  <link rel="stylesheet" href="styles.css">
  <link href="index_files/libs/revealjs/plugin/quarto-line-highlight/line-highlight.css" rel="stylesheet">
  <link href="index_files/libs/revealjs/plugin/reveal-menu/menu.css" rel="stylesheet">
  <link href="index_files/libs/revealjs/plugin/reveal-menu/quarto-menu.css" rel="stylesheet">
  <link href="index_files/libs/revealjs/plugin/reveal-chalkboard/font-awesome/css/all.css" rel="stylesheet">
  <link href="index_files/libs/revealjs/plugin/reveal-chalkboard/style.css" rel="stylesheet">
  <link href="index_files/libs/revealjs/plugin/quarto-support/footer.css" rel="stylesheet">
  <style type="text/css">

  .callout {
    margin-top: 1em;
    margin-bottom: 1em;  
    border-radius: .25rem;
  }

  .callout.callout-style-simple { 
    padding: 0em 0.5em;
    border-left: solid #acacac .3rem;
    border-right: solid 1px silver;
    border-top: solid 1px silver;
    border-bottom: solid 1px silver;
    display: flex;
  }

  .callout.callout-style-default {
    border-left: solid #acacac .3rem;
    border-right: solid 1px silver;
    border-top: solid 1px silver;
    border-bottom: solid 1px silver;
  }

  .callout .callout-body-container {
    flex-grow: 1;
  }

  .callout.callout-style-simple .callout-body {
    font-size: 1rem;
    font-weight: 400;
  }

  .callout.callout-style-default .callout-body {
    font-size: 0.9rem;
    font-weight: 400;
  }

  .callout.callout-titled.callout-style-simple .callout-body {
    margin-top: 0.2em;
  }

  .callout:not(.callout-titled) .callout-body {
      display: flex;
  }

  .callout:not(.no-icon).callout-titled.callout-style-simple .callout-content {
    padding-left: 1.6em;
  }

  .callout.callout-titled .callout-header {
    padding-top: 0.2em;
    margin-bottom: -0.2em;
  }

  .callout.callout-titled .callout-title  p {
    margin-top: 0.5em;
    margin-bottom: 0.5em;
  }
    
  .callout.callout-titled.callout-style-simple .callout-content  p {
    margin-top: 0;
  }

  .callout.callout-titled.callout-style-default .callout-content  p {
    margin-top: 0.7em;
  }

  .callout.callout-style-simple div.callout-title {
    border-bottom: none;
    font-size: .9rem;
    font-weight: 600;
    opacity: 75%;
  }

  .callout.callout-style-default  div.callout-title {
    border-bottom: none;
    font-weight: 600;
    opacity: 85%;
    font-size: 0.9rem;
    padding-left: 0.5em;
    padding-right: 0.5em;
  }

  .callout.callout-style-default div.callout-content {
    padding-left: 0.5em;
    padding-right: 0.5em;
  }

  .callout.callout-style-simple .callout-icon::before {
    height: 1rem;
    width: 1rem;
    display: inline-block;
    content: "";
    background-repeat: no-repeat;
    background-size: 1rem 1rem;
  }

  .callout.callout-style-default .callout-icon::before {
    height: 0.9rem;
    width: 0.9rem;
    display: inline-block;
    content: "";
    background-repeat: no-repeat;
    background-size: 0.9rem 0.9rem;
  }

  .callout-title {
    display: flex
  }
    
  .callout-icon::before {
    margin-top: 1rem;
    padding-right: .5rem;
  }

  .callout.no-icon::before {
    display: none !important;
  }

  .callout.callout-titled .callout-body > .callout-content > :last-child {
    padding-bottom: 0.5rem;
    margin-bottom: 0;
  }

  .callout.callout-titled .callout-icon::before {
    margin-top: .5rem;
    padding-right: .5rem;
  }

  .callout:not(.callout-titled) .callout-icon::before {
    margin-top: 1rem;
    padding-right: .5rem;
  }

  /* Callout Types */

  div.callout-note {
    border-left-color: #4582ec !important;
  }

  div.callout-note .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAEU0lEQVRYCcVXTWhcVRQ+586kSUMMxkyaElstCto2SIhitS5Ek8xUKV2poatCcVHtUlFQk8mbaaziwpWgglJwVaquitBOfhQXFlqlzSJpFSpIYyXNjBNiTCck7x2/8/LeNDOZxDuEkgOXe++553zfefee+/OYLOXFk3+1LLrRdiO81yNqZ6K9cG0P3MeFaMIQjXssE8Z1JzLO9ls20MBZX7oG8w9GxB0goaPrW5aNMp1yOZIa7Wv6o2ykpLtmAPs/vrG14Z+6d4jpbSKuhdcSyq9wGMPXjonwmESXrriLzFGOdDBLB8Y6MNYBu0dRokSygMA/mrun8MGFN3behm6VVAwg4WR3i6FvYK1T7MHo9BK7ydH+1uurECoouk5MPRyVSBrBHMYwVobG2aOXM07sWrn5qgB60rc6mcwIDJtQrnrEr44kmy+UO9r0u9O5/YbkS9juQckLed3DyW2XV/qWBBB3ptvI8EUY3I9p/67OW+g967TNr3Sotn3IuVlfMLVnsBwH4fsnebJvyGm5GeIUA3jljERmrv49SizPYuq+z7c2H/jlGC+Ghhupn/hcapqmcudB9jwJ/3jvnvu6vu5lVzF1fXyZuZZ7U8nRmVzytvT+H3kilYvH09mLWrQdwFSsFEsxFVs5fK7A0g8gMZjbif4ACpKbjv7gNGaD8bUrlk8x+KRflttr22JEMRUbTUwwDQScyzPgedQHZT0xnx7ujw2jfVfExwYHwOsDTjLdJ2ebmeQIlJ7neo41s/DrsL3kl+W2lWvAga0tR3zueGr6GL78M3ifH0rGXrBC2aAR8uYcIA5gwV8zIE8onoh8u0Fca/ciF7j1uOzEnqcIm59sEXoGc0+z6+H45V1CvAvHcD7THztu669cnp+L0okAeIc6zjbM/24LgGM1gZk7jnRu1aQWoU9sfUOuhrmtaPIO3YY1KLLWZaEO5TKUbMY5zx8W9UJ6elpLwKXbsaZ4EFl7B4bMtDv0iRipKoDQT2sNQI9b1utXFdYisi+wzZ/ri/1m7QfDgEuvgUUEIJPq3DhX/5DWNqIXDOweC2wvIR90Oq3lDpdMIgD2r0dXvGdsEW5H6x6HLRJYU7C69VefO1x8Gde1ZFSJLfWS1jbCnhtOPxmpfv2LXOA2Xk2tvnwKKPFuZ/oRmwBwqRQDcKNeVQkYcOjtWVBuM/JuYw5b6isojIkYxyYAFn5K7ZBF10fea52y8QltAg6jnMqNHFBmGkQ1j+U43HMi2xMar1Nv0zGsf1s8nUsmUtPOOrbFIR8bHFDMB5zL13Gmr/kGlCkUzedTzzmzsaJXhYawnA3UmARpiYj5ooJZiUoxFRtK3X6pgNPv+IZVPcnwbOl6f+aBaO1CNvPW9n9LmCp01nuSaTRF2YxHqZ8DYQT6WsXT+RD6eUztwYLZ8rM+rcPxamv1VQzFUkzFXvkiVrySGQgJNvXHJAxiU3/NwiC03rSf05VBaPtu/Z7/B8Yn/w7eguloAAAAAElFTkSuQmCC');
  }

  div.callout-note.callout-style-default .callout-title {
    background-color: #dae6fb
  }

  div.callout-important {
    border-left-color: #d9534f !important;
  }

  div.callout-important .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAEKklEQVRYCcVXTWhcVRS+575MJym48A+hSRFr00ySRQhURRfd2HYjk2SSTokuBCkU2o0LoSKKraKIBTcuFCoidGFD08nkBzdREbpQ1EDNIv8qSGMFUboImMSZd4/f9zJv8ibJMC8xJQfO3HPPPef7zrvvvnvviIkpC9nsw0UttFunbUhpFzFtarSd6WJkStVMw5xyVqYTvkwfzuf/5FgtkVoB0729j1rjXwThS7Vio+Mo6DNnvLfahoZ+i/o32lULuJ3NNiz7q6+pyAUkJaFF6JwaM2lUJlV0MlnQn5aTRbEu0SEqHUa0A4AdiGuB1kFXRfVyg5d87+Dg4DL6m2TLAub60ilj7A1Ec4odSAc8X95sHh7+ZRPCFo6Fnp7HfU/fBng/hi10CjCnWnJjsxvDNxWw0NfV6Rv5GgP3I3jGWXumdTD/3cbEOP2ZbOZp69yniG3FQ9z1jD7bnBu9Fc2tKGC2q+uAJOQHBDRiZX1x36o7fWBs7J9ownbtO+n0/qWkvW7UPIfc37WgT6ZGR++EOJyeQDSb9UB+DZ1G6DdLDzyS+b/kBCYGsYgJbSQHuThGKRcw5xdeQf8YdNHsc6ePXrlSYMBuSIAFTGAtQo+VuALo4BX83N190NWZWbynBjhOHsmNfFWLeL6v+ynsA58zDvvAC8j5PkbOcXCMg2PZFk3q8MjI7WAG/Dp9AwP7jdGBOOQkAvlFUB+irtm16I1Zw9YBcpGTGXYmk3kQIC/Cds55l+iMI3jqhjAuaoe+am2Jw5GT3Nbz3CkE12NavmzN5+erJW7046n/CH1RO/RVa8lBLozXk9uqykkGAyRXLWlLv5jyp4RFsG5vGVzpDLnIjTWgnRy2Rr+tDKvRc7Y8AyZq10jj8DqXdnIRNtFZb+t/ZRtXcDiVnzpqx8mPcDWxgARUqx0W1QB9MeUZiNrV4qP+Ehc+BpNgATsTX8ozYKL2NtFYAHc84fG7ndxUPr+AR/iQSns7uSUufAymwDOb2+NjK27lEFocm/EE2WpyIy/Hi66MWuMKJn8RvxIcj87IM5Vh9663ziW36kR0HNenXuxmfaD8JC7tfKbrhFr7LiZCrMjrzTeGx+PmkosrkNzW94ObzwocJ7A1HokLolY+AvkTiD/q1H0cN48c5EL8Crkttsa/AXQVDmutfyku0E7jShx49XqV3MFK8IryDhYVbj7Sj2P2eBxwcXoe8T8idsKKPRcnZw1b+slFTubwUwhktrfnAt7J++jwQtLZcm3sr9LQrjRzz6cfMv9aLvgmnAGvpoaGLxM4mAEaLV7iAzQ3oU0IvD5x9ix3yF2RAAuYAOO2f7PEFWCXZ4C9Pb2UsgDeVnFSpbFK7/IWu7TPTvBqzbGdCHOJQSxiEjt6IyZmxQyEJHv6xyQsYk//moVFsN2zP6fRImjfq7/n/wFDguUQFNEwugAAAABJRU5ErkJggg==');
  }

  div.callout-important.callout-style-default .callout-title {
    background-color: #f7dddc
  }

  div.callout-warning {
    border-left-color: #f0ad4e !important;
  }

  div.callout-warning .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAETklEQVRYCeVWW2gcVRg+58yaTUnizqbipZeX4uWhBEniBaoUX1Ioze52t7sRq6APio9V9MEaoWlVsFasRq0gltaAPuxms8lu0gcviE/FFOstVbSIxgcv6SU7EZqmdc7v9+9mJtNks51NTUH84ed889/PP+cmxP+d5FIbMJmNbpREu4WUkiTtCicKny0l1pIKmBzovF2S+hIJHX8iEu3hZJ5lNZGqyRrGSIQpq15AzF28jgpeY6yk6GVdrfFqdrD6Iw+QlB8g0YS2g7dyQmXM/IDhBhT0UCiRf59lfqmmDvzRt6kByV/m4JjtzuaujMUM2c5Z2d6JdKrRb3K2q6mA+oYVz8JnDdKPmmNthzkAk/lN63sYPgevrguc72aZX/L9C6x09GYyxBgCX4NlvyGUHOKELlm5rXeR1kchuChJt4SSwyddZRXgvwMGvYo4QSlk3/zkHD8UHxwVJA6zjZZqP8v8kK8OWLnIZtLyCAJagYC4rTGW/9Pqj92N/c+LUaAj27movwbi19tk/whRCIE7Q9vyI6yvRpftAKVTdUjOW40X3h5OXsKCdmFcx0xlLJoSuQngnrJe7Kcjm4OMq9FlC7CMmScQANuNvjfP3PjGXDBaUQmbp296S5L4DrpbrHN1T87ZVEZVCzg1FF0Ft+dKrlLukI+/c9ENo+TvlTDbYFvuKPtQ9+l052rXrgKoWkDAFnvh0wTOmYn8R5f4k/jN/fZiCM1tQx9jQQ4ANhqG4hiL0qIFTGViG9DKB7GYzgubnpofgYRwO+DFjh0Zin2m4b/97EDkXkc+f6xYAPX0KK2I/7fUQuwzuwo/L3AkcjugPNixC8cHf0FyPjWlItmLxWw4Ou9YsQCr5fijMGoD/zpdRy95HRysyXA74MWOnscpO4j2y3HAVisw85hX5+AFBRSHt4ShfLFkIMXTqyKFc46xdzQM6XbAi702a7sy04J0+feReMFKp5q9esYLCqAZYw/k14E/xcLLsFElaornTuJB0svMuJINy8xkIYuL+xPAlWRceH6+HX7THJ0djLUom46zREu7tTkxwmf/FdOZ/sh6Q8qvEAiHpm4PJ4a/doJe0gH1t+aHRgCzOvBvJedEK5OFE5jpm4AGP2a8Dxe3gGJ/pAutug9Gp6he92CsSsWBaEcxGx0FHytmIpuqGkOpldqNYQK8cSoXvd+xLxXADw0kf6UkJNFtdo5MOgaLjiQOQHcn+A6h5NuL2s0qsC2LOM75PcF3yr5STuBSAcGG+meA14K/CI21HcS4LBT6tv0QAh8Dr5l93AhZzG5ZJ4VxAqdZUEl9z7WJ4aN+svMvwHHL21UKTd1mqvChH7/Za5xzXBBKrUcB0TQ+Ulgkfbi/H/YT5EptrGzsEK7tR1B7ln9BBwckYfMiuSqklSznIuoIIOM42MQO+QnduCoFCI0bpkzjCjddHPN/F+2Yu+sd9bKNpVwHhbS3LluK/0zgfwD0xYI5dXuzlQAAAABJRU5ErkJggg==');
  }

  div.callout-warning.callout-style-default .callout-title {
    background-color: #fcefdc
  }

  div.callout-tip {
    border-left-color: #02b875 !important;
  }

  div.callout-tip .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAADr0lEQVRYCe1XTWgTQRj9ZjZV8a9SPIkKgj8I1bMHsUWrqYLVg4Ue6v9BwZOxSYsIerFao7UiUryIqJcqgtpimhbBXoSCVxUFe9CTiogUrUp2Pt+3aUI2u5vdNh4dmMzOzHvvezuz8xNFM0mjnbXaNu1MvFWRXkXEyE6aYOYJpdW4IXuA4r0fo8qqSMDBU0v1HJUgVieAXxzCsdE/YJTdFcVIZQNMyhruOMJKXYFoLfIfIvVIMWdsrd+Rpd86ZmyzzjJmLStqRn0v8lzkb4rVIXvnpScOJuAn2ACC65FkPzEdEy4TPWRLJ2h7z4cArXzzaOdKlbOvKKX25Wl00jSnrwVxAg3o4dRxhO13RBSdNvH0xSARv3adTXbBdTf64IWO2vH0LT+cv4GR1DJt+DUItaQogeBX/chhbTBxEiZ6gftlDNXTrvT7co4ub5A6gp9HIcHvzTa46OS5fBeP87Qm0fQkr4FsYgVQ7Qg+ZayaDg9jhg1GkWj8RG6lkeSacrrHgDaxdoBiZPg+NXV/KifMuB6//JmYH4CntVEHy/keA6x4h4CU5oFy8GzrBS18cLJMXcljAKB6INjWsRcuZBWVaS3GDrqB7rdapVIeA+isQ57Eev9eCqzqOa81CY05VLd6SamW2wA2H3SiTbnbSxmzfp7WtKZkqy4mdyAlGx7ennghYf8voqp9cLSgKdqNfa6RdRsAAkPwRuJZNbpByn+RrJi1RXTwdi8RQF6ymDwGMAtZ6TVE+4uoKh+MYkcLsT0Hk8eAienbiGdjJHZTpmNjlbFJNKDVAp2fJlYju6IreQxQ08UJDNYdoLSl6AadO+fFuCQqVMB1NJwPm69T04Wv5WhfcWyfXQB+wXRs1pt+nCknRa0LVzSA/2B+a9+zQJadb7IyyV24YAxKp2Jqs3emZTuNnKxsah+uabKbMk7CbTgJx/zIgQYErIeTKRQ9yD9wxVof5YolPHqaWo7TD6tJlh7jQnK5z2n3+fGdggIOx2kaa2YI9QWarc5Ce1ipNWMKeSG4DysFF52KBmTNMmn5HqCFkwy34rDg05gDwgH3bBi+sgFhN/e8QvRn8kbamCOhgrZ9GJhFDgfcMHzFb6BAtjKpFhzTjwv1KCVuxHvCbsSiEz4CANnj84cwHdFXAbAOJ4LTSAawGWFn5tDhLMYz6nWeU2wJfIhmIJBefcd/A5FWQWGgrWzyORZ3Q6HuV+Jf0Bj+BTX69fm1zWgK7By1YTXchFDORywnfQ7GpzOo6S+qECrsx2ifVQAAAABJRU5ErkJggg==');
  }

  div.callout-tip.callout-style-default .callout-title {
    background-color: #ccf1e3
  }

  div.callout-caution {
    border-left-color: #fd7e14 !important;
  }

  div.callout-caution .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAACV0lEQVRYCdVWzWoUQRCuqp2ICBLJXgITZL1EfQDBW/bkzUMUD7klD+ATSHBEfAIfQO+iXsWDxJsHL96EHAwhgzlkg8nBg25XWb0zIb0zs9muYYWkoKeru+vn664fBqElyZNuyh167NXJ8Ut8McjbmEraKHkd7uAnAFku+VWdb3reSmRV8PKSLfZ0Gjn3a6Xlcq9YGb6tADjn+lUfTXtVmaZ1KwBIvFI11rRXlWlatwIAAv2asaa9mlB9wwygiDX26qaw1yYPzFXg2N1GgG0FMF8Oj+VIx7E/03lHx8UhvYyNZLN7BwSPgekXXLribw7w5/c8EF+DBK5idvDVYtEEwMeYefjjLAdEyQ3M9nfOkgnPTEkYU+sxMq0BxNR6jExrAI31H1rzvLEfRIdgcv1XEdj6QTQAS2wtstEALLG1yEZ3QhH6oDX7ExBSFEkFINXH98NTrme5IOaaA7kIfiu2L8A3qhH9zRbukdCqdsA98TdElyeMe5BI8Rs2xHRIsoTSSVFfCFCWGPn9XHb4cdobRIWABNf0add9jakDjQJpJ1bTXOJXnnRXHRf+dNL1ZV1MBRCXhMbaHqGI1JkKIL7+i8uffuP6wVQAzO7+qVEbF6NbS0LJureYcWXUUhH66nLR5rYmva+2tjRFtojkM2aD76HEGAD3tPtKM309FJg5j/K682ywcWJ3PASCcycH/22u+Bh7Aa0ehM2Fu4z0SAE81HF9RkB21c5bEn4Dzw+/qNOyXr3DCTQDMBOdhi4nAgiFDGCinIa2owCEChUwD8qzd03PG+qdW/4fDzjUMcE1ZpIAAAAASUVORK5CYII=');
  }

  div.callout-caution.callout-style-default .callout-title {
    background-color: #ffe5d0
  }

  </style>
  <style type="text/css">
    .reveal div.sourceCode {
      margin: 0;
      overflow: auto;
    }
    .reveal div.hanging-indent {
      margin-left: 1em;
      text-indent: -1em;
    }
    .reveal .slide:not(.center) {
      height: 100%;
      overflow-y: auto;
    }
    .reveal .slide.scrollable {
      overflow-y: auto;
    }
    .reveal .footnotes {
      height: 100%;
      overflow-y: auto;
    }
    .reveal .slide .absolute {
      position: absolute;
      display: block;
    }
    .reveal .footnotes ol {
      counter-reset: ol;
      list-style-type: none; 
      margin-left: 0;
    }
    .reveal .footnotes ol li:before {
      counter-increment: ol;
      content: counter(ol) ". "; 
    }
    .reveal .footnotes ol li > p:first-child {
      display: inline-block;
    }
    .reveal .slide ul,
    .reveal .slide ol {
      margin-bottom: 0.5em;
    }
    .reveal .slide ul li,
    .reveal .slide ol li {
      margin-top: 0.4em;
      margin-bottom: 0.2em;
    }
    .reveal .slide ul[role="tablist"] li {
      margin-bottom: 0;
    }
    .reveal .slide ul li > *:first-child,
    .reveal .slide ol li > *:first-child {
      margin-block-start: 0;
    }
    .reveal .slide ul li > *:last-child,
    .reveal .slide ol li > *:last-child {
      margin-block-end: 0;
    }
    .reveal .slide .columns:nth-child(3) {
      margin-block-start: 0.8em;
    }
    .reveal blockquote {
      box-shadow: none;
    }
    .reveal .tippy-content>* {
      margin-top: 0.2em;
      margin-bottom: 0.7em;
    }
    .reveal .tippy-content>*:last-child {
      margin-bottom: 0.2em;
    }
    .reveal .slide > img.stretch.quarto-figure-center,
    .reveal .slide > img.r-stretch.quarto-figure-center {
      display: block;
      margin-left: auto;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-left,
    .reveal .slide > img.r-stretch.quarto-figure-left  {
      display: block;
      margin-left: 0;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-right,
    .reveal .slide > img.r-stretch.quarto-figure-right  {
      display: block;
      margin-left: auto;
      margin-right: 0; 
    }
  </style>
  <link href="index_files/libs/htmltools-fill-0.5.8.1/fill.css" rel="stylesheet">
  <script src="index_files/libs/htmlwidgets-1.6.4/htmlwidgets.js"></script>
  <link href="index_files/libs/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet">
  <script src="index_files/libs/datatables-binding-0.33/datatables.js"></script>
  <script src="index_files/libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
  <link href="index_files/libs/dt-core-1.13.6/css/jquery.dataTables.min.css" rel="stylesheet">
  <link href="index_files/libs/dt-core-1.13.6/css/jquery.dataTables.extra.css" rel="stylesheet">
  <script src="index_files/libs/dt-core-1.13.6/js/jquery.dataTables.min.js"></script>
  <link href="index_files/libs/crosstalk-1.2.1/css/crosstalk.min.css" rel="stylesheet">
  <script src="index_files/libs/crosstalk-1.2.1/js/crosstalk.min.js"></script>
  <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
  <script id="MathJax-script" async="" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>

  <script>
  document.addEventListener('DOMContentLoaded', function() {
      function resizeIframes() {
          var iframes = document.getElementsByClassName('plotlyFrame');
          Array.from(iframes).forEach(function(iframe) {
              // Set the dimensions relative to parent or custom logic
              iframe.style.height = window.innerHeight * 1 + 'px'; // 100% of the viewport height
              iframe.style.width = window.innerWidth * 0.9 + 'px'; // 75% of the viewport width
          });
      }

      // Resize iframes when the window resizes
      window.addEventListener('resize', resizeIframes);

      // Initial resize
      resizeIframes();
  });
  </script>
  <!-- This is also a suitable place for other snippets from JavaScript, CSS, or HTML elements, in order to have them included in the header -->
    <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/monaco-editor@0.45.0/min/vs/editor/editor.main.css">
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.5.1/css/all.min.css" integrity="sha512-DTOQO9RWCH3ppGqcWaEA1BIZOC6xxalwEsw9c2QQeAIftl+Vegovlnee1c9QX4TctnWMn13TZye+giMm8e2LwA==" crossorigin="anonymous" referrerpolicy="no-referrer">
    
  <style type="text/css">
  .monaco-editor pre {
    background-color: unset !important;
  }

  .qwebr-editor-toolbar {
    width: 100%;
    display: flex;
    justify-content: space-between;
    box-sizing: border-box;
  }

  .qwebr-editor-toolbar-left-buttons, .qwebr-editor-toolbar-right-buttons {
    display: flex;
  }

  .qwebr-non-interactive-loading-container.qwebr-cell-needs-evaluation, .qwebr-non-interactive-loading-container.qwebr-cell-evaluated {
    justify-content: center;
    display: flex;
    background-color: rgba(250, 250, 250, 0.65);
    border: 1px solid rgba(233, 236, 239, 0.65);
    border-radius: 0.5rem;
    margin-top: 15px;
    margin-bottom: 15px;
  }

  .qwebr-r-project-logo {
    color: #2767B0; /* R Project's blue color */
  }

  .qwebr-icon-status-spinner {
    color: #7894c4;
  }

  .qwebr-icon-run-code {
    color: #0d9c29
  }

  .qwebr-output-code-stdout {
    color: #111;
  }

  .qwebr-output-code-stderr {
    color: #db4133;
  }

  .qwebr-editor {
    border: 1px solid #EEEEEE;
  }

  .qwebr-editor-toolbar {
    background-color: #EEEEEE;
    padding: 0.2rem 0.5rem;
  }

  .qwebr-button {
    background-color: #EEEEEE;
    display: inline-block;
    font-weight: 400;
    line-height: 1;
    text-decoration: none;
    text-align: center;
    color: #000;
    border-color: #dee2e6;
    border: 1px solid rgba(0,0,0,0);
    padding: 0.375rem 0.75rem;
    font-size: .9rem;
    border-radius: 0.25rem;
    transition: color .15s ease-in-out,background-color .15s ease-in-out,border-color .15s ease-in-out,box-shadow .15s ease-in-out;
  }

  .qwebr-button:hover {
    color: #000;
    background-color: #d9dce0;
    border-color: #c8ccd0;
  }

  .qwebr-button:disabled,.qwebr-button.disabled,fieldset:disabled .qwebr-button {
    pointer-events: none;
    opacity: .65
  }

  .qwebr-button-reset {
    color: #696969; /*#4682b4;*/
  }

  .qwebr-button-copy {
    color: #696969;
  }


  /* Custom styling for RevealJS Presentations*/

  /* Reset the style of the interactive area */
  .reveal div.qwebr-interactive-area {
    display: block;
    box-shadow: none;
    max-width: 100%;
    max-height: 100%;
    margin: 0;
    padding: 0;
  } 

  /* Provide space to entries */
  .reveal div.qwebr-output-code-area pre div {
    margin: 1px 2px 1px 10px;
  }

  /* Collapse the inside code tags to avoid extra space between line outputs */
  .reveal pre div code.qwebr-output-code-stdout, .reveal pre div code.qwebr-output-code-stderr {
    padding: 0;
    display: contents;
  }

  .reveal pre div code.qwebr-output-code-stdout {
    color: #111;
  }

  .reveal pre div code.qwebr-output-code-stderr {
    color: #db4133;
  }


  /* Create a border around console and output (does not effect graphs) */
  .reveal div.qwebr-console-area {
    border: 1px solid #EEEEEE;
    box-shadow: 2px 2px 10px #EEEEEE;
  }

  /* Cap output height and allow text to scroll */
  /* TODO: Is there a better way to fit contents/max it parallel to the monaco editor size? */
  .reveal div.qwebr-output-code-area pre {
    max-height: 400px;
    overflow: scroll;
  }

  </style>
  <script type="module">
  // Document level settings ----

  // Determine if we need to install R packages
  globalThis.qwebrInstallRPackagesList = ['tibble', 'dplyr', 'feather'];

  // Specify possible locations to search for the repository
  globalThis.qwebrPackageRepoURLS = ['https://repo.r-wasm.org/'];

  // Check to see if we have an empty array, if we do set to skip the installation.
  globalThis.qwebrSetupRPackages = !(qwebrInstallRPackagesList.indexOf("") !== -1);
  globalThis.qwebrAutoloadRPackages = true;

  // Display a startup message?
  globalThis.qwebrShowStartupMessage = false;
  globalThis.qwebrShowHeaderMessage = false;

  // Describe the webR settings that should be used
  globalThis.qwebrCustomizedWebROptions = {
    "baseURL": "https://webr.r-wasm.org/v0.2.2/",
    "serviceWorkerUrl": "",
    "homedir": "/home/web_user", 
    "channelType": "ChannelType.Automatic"
  };

  // Store cell data
  globalThis.qwebrCellDetails = [{"id":1,"options":{"comment":"","classes":"","message":"true","warning":"true","context":"setup","out-width":"700px","dpi":72,"read-only":"false","label":"unnamed-chunk-1","results":"markup","fig-height":5,"output":"true","fig-cap":"","autorun":"true","out-height":"","fig-width":7},"code":"options(max.print = 10)\n# Download a dataset\n# url <- \"https://raw.githubusercontent.com/DHintz137/Embedding_Presentation/main/raw-data/variant_data_subset.feather\"\n# download.file(url, \"variant_data_subset.feather\")\n# Whole_genome_covid_seqs <- feather::read_feather(\"raw-data/variant_data_subset.feather\")"},{"id":2,"options":{"comment":"","classes":"","message":"true","warning":"true","context":"setup","out-width":"700px","dpi":72,"read-only":"false","label":"unnamed-chunk-2","results":"markup","fig-height":5,"output":"true","fig-cap":"","autorun":"true","out-height":"","fig-width":7},"code":"\noptions(max.print = 15)\n# Download a dataset\nurl <- \"https://raw.githubusercontent.com/DHintz137/Embedding_Presentation/main/raw-data/King_Man_Woman_Queen_pred.csv\"\ndownload.file(url, \"King_Man_Woman_Queen_pred.csv\")\nembedding_raw <- read.csv(\"King_Man_Woman_Queen_pred.csv\")\nembedding <- t(data.frame(embedding_raw[,-1] ,row.names = embedding_raw[,1]))\ncolnames(embedding) <- c(\"King\", \"Man\", \"Woman\", \"Queen\")\n# embedding <- as_tibble(embedding)\n\nprint_tidy <- function(x, ...) {\n  if (is.matrix(x) || length(dim(x)) > 1) {\n    # If x is a matrix or array, convert each column into a tibble column\n    out <- as_tibble(as.data.frame(x))\n  } else if (is.vector(x)) {\n    # If x is a vector, convert it into a single column tibble\n    out <- tibble(value = x)\n  } else {\n    # Fallback for other data types\n    out <- as_tibble(x)\n  }\n  print(out, ...)\n}"},{"id":3,"options":{"comment":"","classes":"small-code","out-width":"700px","label":"unnamed-chunk-3","results":"markup","context":"interactive","message":"true","class-source":"small-code","warning":"true","dpi":72,"read-only":"false","fig-cap":"","fig-height":5,"fig-width":7,"autorun":"false","output":"true","out-height":""},"code":"print_tidy(embedding)\nprint_tidy(embedding[ ,\"King\"] - embedding[ , \"Man\"] + embedding[ , \"Woman\"])\nprint_tidy(t(embedding[ , \"Man\"]) %*% embedding[ , \"Woman\"])"}];

  </script>
  <script type="module">
  // Declare startupMessageQWebR globally
  globalThis.qwebrStartupMessage = document.createElement("p");


  // Function to set the button text
  globalThis.qwebrSetInteractiveButtonState = function(buttonText, enableCodeButton = true) {
    document.querySelectorAll(".qwebr-button-run").forEach((btn) => {
      btn.innerHTML = buttonText;
      btn.disabled = !enableCodeButton;
    });
  }

  // Function to update the status message in non-interactive cells
  globalThis.qwebrUpdateStatusMessage = function(message) {
    document.querySelectorAll(".qwebr-status-text.qwebr-cell-needs-evaluation").forEach((elem) => {
      elem.innerText = message;
    });
  }

  // Function to update the status message
  globalThis.qwebrUpdateStatusHeader = function(message) {
    qwebrStartupMessage.innerHTML = `
      <i class="fa-solid fa-spinner fa-spin qwebr-icon-status-spinner"></i>
      <span>${message}</span>`;
  }

  // Function that attaches the document status message and diagnostics
  function displayStartupMessage(showStartupMessage, showHeaderMessage) {
    if (!showStartupMessage) {
      return;
    }

    // Get references to header elements
    const headerHTML = document.getElementById("title-block-header");
    const headerRevealJS = document.getElementById("title-slide");

    // Create the outermost div element for metadata
    const quartoTitleMeta = document.createElement("div");
    quartoTitleMeta.classList.add("quarto-title-meta");

    // Create the first inner div element
    const firstInnerDiv = document.createElement("div");
    firstInnerDiv.setAttribute("id", "qwebr-status-message-area");

    // Create the second inner div element for "WebR Status" heading and contents
    const secondInnerDiv = document.createElement("div");
    secondInnerDiv.setAttribute("id", "qwebr-status-message-title");
    secondInnerDiv.classList.add("quarto-title-meta-heading");
    secondInnerDiv.innerText = "WebR Status";

    // Create another inner div for contents
    const secondInnerDivContents = document.createElement("div");
    secondInnerDivContents.setAttribute("id", "qwebr-status-message-body");
    secondInnerDivContents.classList.add("quarto-title-meta-contents");

    // Describe the WebR state
    qwebrStartupMessage.innerText = "🟡 Loading...";
    qwebrStartupMessage.setAttribute("id", "qwebr-status-message-text");
    // Add `aria-live` to auto-announce the startup status to screen readers
    qwebrStartupMessage.setAttribute("aria-live", "assertive");

    // Append the startup message to the contents
    secondInnerDivContents.appendChild(qwebrStartupMessage);

    // Add a status indicator for COOP and COEP Headers if needed
    if (showHeaderMessage) {
      const crossOriginMessage = document.createElement("p");
      crossOriginMessage.innerText = `${crossOriginIsolated ? '🟢' : '🟡'} COOP & COEP Headers`;
      crossOriginMessage.setAttribute("id", "qwebr-coop-coep-header");
      secondInnerDivContents.appendChild(crossOriginMessage);
    }

    // Combine the inner divs and contents
    firstInnerDiv.appendChild(secondInnerDiv);
    firstInnerDiv.appendChild(secondInnerDivContents);
    quartoTitleMeta.appendChild(firstInnerDiv);

    // Determine where to insert the quartoTitleMeta element
    if (headerHTML || headerRevealJS) {
      // Append to the existing "title-block-header" element or "title-slide" div
      (headerHTML || headerRevealJS).appendChild(quartoTitleMeta);
    } else {
      // If neither headerHTML nor headerRevealJS is found, insert after "webr-monaco-editor-init" script
      const monacoScript = document.getElementById("qwebr-monaco-editor-init");
      const header = document.createElement("header");
      header.setAttribute("id", "title-block-header");
      header.appendChild(quartoTitleMeta);
      monacoScript.after(header);
    }
  }

  displayStartupMessage(qwebrShowStartupMessage, qwebrShowHeaderMessage);
  </script>
  <script type="module">
  // Supported Evaluation Types for Context
  globalThis.EvalTypes = Object.freeze({
    Interactive: 'interactive',
    Setup: 'setup',
    Output: 'output',
  });

  // Function that dispatches the creation request
  globalThis.qwebrCreateHTMLElement = function (
    cellData
  ) {

    // Extract key components
    const evalType = cellData.options.context;
    const qwebrCounter = cellData.id;

    // We make an assumption that insertion points are defined by the Lua filter as:
    // qwebr-insertion-location-{qwebrCounter} 
    const elementLocator = document.getElementById(`qwebr-insertion-location-${qwebrCounter}`);

    // Figure out the routine to use to insert the element.
    let qwebrElement;
    switch ( evalType ) {
      case EvalTypes.Interactive:
        qwebrElement = qwebrCreateInteractiveElement(qwebrCounter, cellData.options);
        break;
      case EvalTypes.Output: 
        qwebrElement = qwebrCreateNonInteractiveOutputElement(qwebrCounter, cellData.options);
        break;
      case EvalTypes.Setup: 
        qwebrElement = qwebrCreateNonInteractiveSetupElement(qwebrCounter, cellData.options);
        break;
      default: 
        qwebrElement = document.createElement('div');
        qwebrElement.textContent = 'Error creating `quarto-webr` element';
    }

    // Insert the dynamically generated object at the document location.
    elementLocator.appendChild(qwebrElement);
  };

  // Function that setups the interactive element creation
  globalThis.qwebrCreateInteractiveElement = function (qwebrCounter, qwebrOptions) {

    // Create main div element
    var mainDiv = document.createElement('div');
    mainDiv.id = 'qwebr-interactive-area-' + qwebrCounter;
    mainDiv.className = `qwebr-interactive-area`;
    if (qwebrOptions.classes) {
      mainDiv.className += " " + qwebrOptions.classes
    }

    // Add a unique cell identifier that users can customize
    if (qwebrOptions.label) {
      mainDiv.setAttribute('data-id', qwebrOptions.label);
    }

    // Create toolbar div
    var toolbarDiv = document.createElement('div');
    toolbarDiv.className = 'qwebr-editor-toolbar';
    toolbarDiv.id = 'qwebr-editor-toolbar-' + qwebrCounter;

    // Create a div to hold the left buttons
    var leftButtonsDiv = document.createElement('div');
    leftButtonsDiv.className = 'qwebr-editor-toolbar-left-buttons';

    // Create a div to hold the right buttons
    var rightButtonsDiv = document.createElement('div');
    rightButtonsDiv.className = 'qwebr-editor-toolbar-right-buttons';

    // Create Run Code button
    var runCodeButton = document.createElement('button');
    runCodeButton.className = 'btn btn-default qwebr-button qwebr-button-run';
    runCodeButton.disabled = true;
    runCodeButton.type = 'button';
    runCodeButton.id = 'qwebr-button-run-' + qwebrCounter;
    runCodeButton.textContent = '🟡 Loading webR...';
    runCodeButton.title = `Run code (Shift + Enter)`;

    // Append buttons to the leftButtonsDiv
    leftButtonsDiv.appendChild(runCodeButton);

    // Create Reset button
    var resetButton = document.createElement('button');
    resetButton.className = 'btn btn-light btn-xs qwebr-button qwebr-button-reset';
    resetButton.type = 'button';
    resetButton.id = 'qwebr-button-reset-' + qwebrCounter;
    resetButton.title = 'Start over';
    resetButton.innerHTML = '<i class="fa-solid fa-arrows-rotate"></i>';

    // Create Copy button
    var copyButton = document.createElement('button');
    copyButton.className = 'btn btn-light btn-xs qwebr-button qwebr-button-copy';
    copyButton.type = 'button';
    copyButton.id = 'qwebr-button-copy-' + qwebrCounter;
    copyButton.title = 'Copy code';
    copyButton.innerHTML = '<i class="fa-regular fa-copy"></i>';

    // Append buttons to the rightButtonsDiv
    rightButtonsDiv.appendChild(resetButton);
    rightButtonsDiv.appendChild(copyButton);

    // Create console area div
    var consoleAreaDiv = document.createElement('div');
    consoleAreaDiv.id = 'qwebr-console-area-' + qwebrCounter;
    consoleAreaDiv.className = 'qwebr-console-area';

    // Create editor div
    var editorDiv = document.createElement('div');
    editorDiv.id = 'qwebr-editor-' + qwebrCounter;
    editorDiv.className = 'qwebr-editor';

    // Create output code area div
    var outputCodeAreaDiv = document.createElement('div');
    outputCodeAreaDiv.id = 'qwebr-output-code-area-' + qwebrCounter;
    outputCodeAreaDiv.className = 'qwebr-output-code-area';
    outputCodeAreaDiv.setAttribute('aria-live', 'assertive');

    // Create pre element inside output code area
    var preElement = document.createElement('pre');
    preElement.style.visibility = 'hidden';
    outputCodeAreaDiv.appendChild(preElement);

    // Create output graph area div
    var outputGraphAreaDiv = document.createElement('div');
    outputGraphAreaDiv.id = 'qwebr-output-graph-area-' + qwebrCounter;
    outputGraphAreaDiv.className = 'qwebr-output-graph-area';

    // Append buttons to the toolbar
    toolbarDiv.appendChild(leftButtonsDiv);
    toolbarDiv.appendChild(rightButtonsDiv);

    // Append all elements to the main div
    mainDiv.appendChild(toolbarDiv);
    consoleAreaDiv.appendChild(editorDiv);
    consoleAreaDiv.appendChild(outputCodeAreaDiv);
    mainDiv.appendChild(consoleAreaDiv);
    mainDiv.appendChild(outputGraphAreaDiv);

    return mainDiv;
  }

  // Function that adds output structure for non-interactive output
  globalThis.qwebrCreateNonInteractiveOutputElement = function(qwebrCounter, qwebrOptions) {
    // Create main div element
    var mainDiv = document.createElement('div');
    mainDiv.id = 'qwebr-noninteractive-area-' + qwebrCounter;
    mainDiv.className = `qwebr-noninteractive-area`;
    if (qwebrOptions.classes) {
      mainDiv.className += " " + qwebrOptions.classes
    }
    
    // Add a unique cell identifier that users can customize
    if (qwebrOptions.label) {
      mainDiv.setAttribute('data-id', qwebrOptions.label);
    }
    
    // Create a status container div
    var statusContainer = createLoadingContainer(qwebrCounter);

    // Create output code area div
    var outputCodeAreaDiv = document.createElement('div');
    outputCodeAreaDiv.id = 'qwebr-output-code-area-' + qwebrCounter;
    outputCodeAreaDiv.className = 'qwebr-output-code-area';
    outputCodeAreaDiv.setAttribute('aria-live', 'assertive');

    // Create pre element inside output code area
    var preElement = document.createElement('pre');
    preElement.style.visibility = 'hidden';
    outputCodeAreaDiv.appendChild(preElement);

    // Create output graph area div
    var outputGraphAreaDiv = document.createElement('div');
    outputGraphAreaDiv.id = 'qwebr-output-graph-area-' + qwebrCounter;
    outputGraphAreaDiv.className = 'qwebr-output-graph-area';

    // Append all elements to the main div
    mainDiv.appendChild(statusContainer);
    mainDiv.appendChild(outputCodeAreaDiv);
    mainDiv.appendChild(outputGraphAreaDiv);

    return mainDiv;
  };

  // Function that adds a stub in the page to indicate a setup cell was used.
  globalThis.qwebrCreateNonInteractiveSetupElement = function(qwebrCounter, qwebrOptions) {
    // Create main div element
    var mainDiv = document.createElement('div');
    mainDiv.id = `qwebr-noninteractive-setup-area-${qwebrCounter}`;
    mainDiv.className = `qwebr-noninteractive-setup-area`;
    if (qwebrOptions.classes) {
      mainDiv.className += " " + qwebrOptions.classes
    }


    // Add a unique cell identifier that users can customize
    if (qwebrOptions.label) {
      mainDiv.setAttribute('data-id', qwebrOptions.label);
    }

    // Create a status container div
    var statusContainer = createLoadingContainer(qwebrCounter);

    // Append status onto the main div
    mainDiv.appendChild(statusContainer);

    return mainDiv;
  }


  // Function to create loading container with specified ID
  globalThis.createLoadingContainer = function(qwebrCounter) {

    // Create a status container
    const container = document.createElement('div');
    container.id = `qwebr-non-interactive-loading-container-${qwebrCounter}`;
    container.className = 'qwebr-non-interactive-loading-container qwebr-cell-needs-evaluation';

    // Create an R project logo to indicate its a code space
    const rProjectIcon = document.createElement('i');
    rProjectIcon.className = 'fa-brands fa-r-project fa-3x qwebr-r-project-logo';

    // Setup a loading icon from font awesome
    const spinnerIcon = document.createElement('i');
    spinnerIcon.className = 'fa-solid fa-spinner fa-spin fa-1x qwebr-icon-status-spinner';

    // Add a section for status text
    const statusText = document.createElement('p');
    statusText.id = `qwebr-status-text-${qwebrCounter}`;
    statusText.className = `qwebr-status-text qwebr-cell-needs-evaluation`;
    statusText.innerText = 'Loading webR...';

    // Incorporate an inner container
    const innerContainer = document.createElement('div');

    // Append elements to the inner container
    innerContainer.appendChild(spinnerIcon);
    innerContainer.appendChild(statusText);

    // Append elements to the main container
    container.appendChild(rProjectIcon);
    container.appendChild(innerContainer);

    return container;
  }
  </script>
  <script type="module">
  // Function to install a single package
  async function qwebrInstallRPackage(packageName) {
    await mainWebR.evalRVoid(`webr::install('${packageName}');`);
  }

  // Function to load a single package
  async function qwebrLoadRPackage(packageName) {
    await mainWebR.evalRVoid(`require('${packageName}', quietly = TRUE)`);
  }

  // Generic function to process R packages
  async function qwebrProcessRPackagesWithStatus(packages, processType, displayStatusMessageUpdate = true) {
    // Switch between contexts
    const messagePrefix = processType === 'install' ? 'Installing' : 'Loading';

    // Modify button state
    qwebrSetInteractiveButtonState(`🟡 ${messagePrefix} package ...`, false);

    // Iterate over packages
    for (let i = 0; i < packages.length; i++) {
      const activePackage = packages[i];
      const formattedMessage = `${messagePrefix} package ${i + 1} out of ${packages.length}: ${activePackage}`;

      // Display the update in header
      if (displayStatusMessageUpdate) {
        qwebrUpdateStatusHeader(formattedMessage);
      }

      // Display the update in non-active areas
      qwebrUpdateStatusMessage(formattedMessage);

      // Run package installation
      if (processType === 'install') {
        await qwebrInstallRPackage(activePackage);
      } else {
        await qwebrLoadRPackage(activePackage);
      }
    }

    // Clean slate
    if (processType === 'load') {
      await mainWebR.flush();
    }
  }

  // Start a timer
  const initializeWebRTimerStart = performance.now();

  // Encase with a dynamic import statement
  globalThis.qwebrInstance = import(qwebrCustomizedWebROptions.baseURL + "webr.mjs").then(
    async ({ WebR, ChannelType }) => {
      // Populate WebR options with defaults or new values based on `webr` meta
      globalThis.mainWebR = new WebR(qwebrCustomizedWebROptions);

      // Initialization WebR
      await mainWebR.init();

      // Setup a shelter
      globalThis.mainWebRCodeShelter = await new mainWebR.Shelter();

      // Setup a pager to allow processing help documentation 
      await mainWebR.evalRVoid('webr::pager_install()'); 

      // Override the existing install.packages() to use webr::install()
      await mainWebR.evalRVoid('webr::shim_install()'); 

      // Specify the repositories to pull from
      // Note: webR does not use the `repos` option, but instead uses `webr_pkg_repos`
      // inside of `install()`. However, other R functions still pull from `repos`.
      await mainWebR.evalRVoid(`
        options(
          webr_pkg_repos = c(${qwebrPackageRepoURLS.map(repoURL => `'${repoURL}'`).join(',')}),
          repos = c(${qwebrPackageRepoURLS.map(repoURL => `'${repoURL}'`).join(',')})
        )
      `);

      // Check to see if any packages need to be installed
      if (qwebrSetupRPackages) {
        // Obtain only a unique list of packages
        const uniqueRPackageList = Array.from(new Set(qwebrInstallRPackagesList));

        // Install R packages one at a time (either silently or with a status update)
        await qwebrProcessRPackagesWithStatus(uniqueRPackageList, 'install', qwebrShowStartupMessage);

        if (qwebrAutoloadRPackages) {
          // Load R packages one at a time (either silently or with a status update)
          await qwebrProcessRPackagesWithStatus(uniqueRPackageList, 'load', qwebrShowStartupMessage);
        }
      }
    }
  );

  // Stop timer
  const initializeWebRTimerEnd = performance.now();

  </script>
  <script type="module">
  // Function to verify a given JavaScript Object is empty
  globalThis.qwebrIsObjectEmpty = function (arr) {
      return Object.keys(arr).length === 0;
  }

  // Global version of the Escape HTML function that converts HTML 
  // characters to their HTML entities.
  globalThis.qwebrEscapeHTMLCharacters = function(unsafe) {
      return unsafe
        .replace(/&/g, "&amp;")
        .replace(/</g, "&lt;")
        .replace(/>/g, "&gt;")
        .replace(/"/g, "&quot;")
        .replace(/'/g, "&#039;");
    };

  // Passthrough results
  globalThis.qwebrIdentity = function(x) {
      return x;
  };

  // Append a comment
  globalThis.qwebrPrefixComment = function(x, comment) {
      return `${comment}${x}`;
  };

  // Function to parse the pager results
  globalThis.qwebrParseTypePager = async function (msg) { 

      // Split out the event data
      const { path, title, deleteFile } = msg.data; 

      // Process the pager data by reading the information from disk
      const paged_data = await mainWebR.FS.readFile(path).then((data) => {
          // Obtain the file content
          let content = new TextDecoder().decode(data);

          // Remove excessive backspace characters until none remain
          while(content.match(/.[\b]/)){
          content = content.replace(/.[\b]/g, '');
          }

          // Returned cleaned data
          return content;
      });

      // Unlink file if needed
      if (deleteFile) { 
          await mainWebR.FS.unlink(path); 
      } 

      // Return extracted data with spaces
      return paged_data;
  } 

  // Function to run the code using webR and parse the output
  globalThis.qwebrComputeEngine = async function(
      codeToRun, 
      elements, 
      options) {

      // Call into the R compute engine that persists within the document scope.
      // To be prepared for all scenarios, the following happens: 
      // 1. We setup a canvas device to write to by making a namespace call into the {webr} package
      // 2. We use values inside of the options array to set the figure size.
      // 3. We capture the output stream information (STDOUT and STERR)
      // 4. While parsing the results, we disable image creation.

      // Create a canvas variable for graphics
      let canvas = undefined;

      // Create a pager variable for help/file contents
      let pager = [];

      // Handle how output is processed
      let showMarkup = options.results === "markup" && options.output !== "asis";
      let processOutput;

      if (showMarkup) {
          processOutput = qwebrEscapeHTMLCharacters;
      } else {
          processOutput = qwebrIdentity;
      }

      // ---- 
      // Convert from Inches to Pixels by using DPI (dots per inch)
      // for bitmap devices (dpi * inches = pixels)
      let fig_width = options["fig-width"] * options["dpi"]
      let fig_height = options["fig-height"] * options["dpi"]

      // Initialize webR
      await mainWebR.init();

      // Setup a webR canvas by making a namespace call into the {webr} package
      await mainWebR.evalRVoid(`webr::canvas(width=${fig_width}, height=${fig_height})`);

      const result = await mainWebRCodeShelter.captureR(codeToRun, {
          withAutoprint: true,
          captureStreams: true,
          captureConditions: false//,
          // env: webR.objs.emptyEnv, // maintain a global environment for webR v0.2.0
      });

      // -----

      // Start attempting to parse the result data
      processResultOutput:try {

          // Stop creating images
          await mainWebR.evalRVoid("dev.off()");
          
          // Avoid running through output processing
          if (options.results === "hide" || options.output === "false") { 
              break processResultOutput; 
          }

          // Merge output streams of STDOUT and STDErr (messages and errors are combined.)
          // Require both `warning` and `message` to be true to display `STDErr`. 
          const out = result.output
          .filter(
              evt => evt.type === "stdout" || 
              ( evt.type === "stderr" && (options.warning === "true" && options.message === "true")) 
          )
          .map((evt, index) => {
              const className = `qwebr-output-code-${evt.type}`;
              const outputResult = qwebrPrefixComment(processOutput(evt.data), options.comment);
              return `<code id="${className}-editor-${elements.id}-result-${index + 1}" class="${className}">${outputResult}</code>`;
          })
          .join("\n");


          // Clean the state
          // We're now able to process both graphics and pager events.
          // As a result, we cannot maintain a true 1-to-1 output order 
          // without individually feeding each line
          const msgs = await mainWebR.flush();

          // Output each image event stored
          msgs.forEach((msg) => {
          // Determine if old canvas can be used or a new canvas is required.
          if (msg.type === 'canvas'){
              // Add image to the current canvas
              if (msg.data.event === 'canvasImage') {
                  canvas.getContext('2d').drawImage(msg.data.image, 0, 0);
              } else if (msg.data.event === 'canvasNewPage') {

                  // Generate a new canvas element
                  canvas = document.createElement("canvas");
                  canvas.setAttribute("width", 2 * fig_width);
                  canvas.setAttribute("height", 2 * fig_height);
                  canvas.style.width = options["out-width"] ? options["out-width"] : `${fig_width}px`;
                  if (options["out-height"]) {
                      canvas.style.height = options["out-height"];
                  }
                  canvas.style.display = "block";
                  canvas.style.margin = "auto";
              }
          } 
          });

          // Use `map` to process the filtered "pager" events asynchronously
          const pager = await Promise.all(
              msgs.filter(msg => msg.type === 'pager').map(
                  async (msg) => {
                      return await qwebrParseTypePager(msg);
                  }
              )
          );

          // Nullify the output area of content
          elements.outputCodeDiv.innerHTML = "";
          elements.outputGraphDiv.innerHTML = "";

          // Design an output object for messages
          const pre = document.createElement("pre");
          if (/\S/.test(out)) {
              // Display results as HTML elements to retain output styling
              const div = document.createElement("div");
              div.innerHTML = out;
              pre.appendChild(div);
          } else {
              // If nothing is present, hide the element.
              pre.style.visibility = "hidden";
          }

          elements.outputCodeDiv.appendChild(pre);

          // Place the graphics on the canvas
          if (canvas) {
              // Create figure element
              const figureElement = document.createElement('figure');

              // Append canvas to figure
              figureElement.appendChild(canvas);

              if (options['fig-cap']) {
                  // Create figcaption element
                  const figcaptionElement = document.createElement('figcaption');
                  figcaptionElement.innerText = options['fig-cap'];
                  // Append figcaption to figure
                  figureElement.appendChild(figcaptionElement);    
              }

              elements.outputGraphDiv.appendChild(figureElement);
          }

          // Display the pager data
          if (pager) {
          // Use the `pre` element to preserve whitespace.
          pager.forEach((paged_data, index) => {
              let pre_pager = document.createElement("pre");
              pre_pager.innerText = paged_data;
              pre_pager.classList.add("qwebr-output-code-pager");
              pre_pager.setAttribute("id", `qwebr-output-code-pager-editor-${elements.id}-result-${index + 1}`);
              elements.outputCodeDiv.appendChild(pre_pager);
          });
          }
      } finally {
          // Clean up the remaining code
          mainWebRCodeShelter.purge();
      }
  }

  // Function to execute the code (accepts code as an argument)
  globalThis.qwebrExecuteCode = async function (
      codeToRun,
      id,
      options = {}) {

      // If options are not passed, we fall back on the bare minimum to handle the computation
      if (qwebrIsObjectEmpty(options)) {
          options = { 
              "context": "interactive", 
              "fig-width": 7, "fig-height": 5, 
              "out-width": "700px", "out-height": "", 
              "dpi": 72,
              "results": "markup", 
              "warning": "true", "message": "true",
          };
      }

      // Next, we access the compute areas values
      const elements = {
          runButton: document.getElementById(`qwebr-button-run-${id}`),
          outputCodeDiv: document.getElementById(`qwebr-output-code-area-${id}`),
          outputGraphDiv: document.getElementById(`qwebr-output-graph-area-${id}`),
          id: id,
      }

      // Disallowing execution of other code cells
      document.querySelectorAll(".qwebr-button-run").forEach((btn) => {
          btn.disabled = true;
      });

      if (options.context == EvalTypes.Interactive) {
          // Emphasize the active code cell
          elements.runButton.innerHTML = '<i class="fa-solid fa-spinner fa-spin qwebr-icon-status-spinner"></i> <span>Run Code</span>';
      }

      // Evaluate the code and parse the output into the document
      await qwebrComputeEngine(codeToRun, elements, options);

      // Switch to allowing execution of code
      document.querySelectorAll(".qwebr-button-run").forEach((btn) => {
          btn.disabled = false;
      });

      if (options.context == EvalTypes.Interactive) {
          // Revert to the initial code cell state
          elements.runButton.innerHTML = '<i class="fa-solid fa-play qwebr-icon-run-code"></i> <span>Run Code</span>';
      }
  }

  </script>
</head>
<body class="quarto-light">
<script src="https://cdn.jsdelivr.net/npm/monaco-editor@0.45.0/min/vs/loader.js"></script>
<script type="module" id="qwebr-monaco-editor-init">

  // Configure the Monaco Editor's loader
  require.config({
    paths: {
      'vs': 'https://cdn.jsdelivr.net/npm/monaco-editor@0.45.0/min/vs'
    }
  });
</script>
<script type="module">
// Global dictionary to store Monaco Editor instances
globalThis.qwebrEditorInstances = {};

// Function that builds and registers a Monaco Editor instance    
globalThis.qwebrCreateMonacoEditorInstance = function (cellData) {

  const initialCode = cellData.code;
  const qwebrCounter = cellData.id;
  const qwebrOptions = cellData.options;

  // Retrieve the previously created document elements
  let runButton = document.getElementById(`qwebr-button-run-${qwebrCounter}`);
  let resetButton = document.getElementById(`qwebr-button-reset-${qwebrCounter}`);
  let copyButton = document.getElementById(`qwebr-button-copy-${qwebrCounter}`);
  let editorDiv = document.getElementById(`qwebr-editor-${qwebrCounter}`);
  
  // Load the Monaco Editor and create an instance
  let editor;
  require(['vs/editor/editor.main'], function () {
    editor = monaco.editor.create(editorDiv, {
      value: initialCode,
      language: 'r',
      theme: 'vs-light',
      automaticLayout: true,           // Works wonderfully with RevealJS
      scrollBeyondLastLine: false,
      minimap: {
        enabled: false
      },
      fontSize: '17.5pt',              // Bootstrap is 1 rem
      renderLineHighlight: "none",     // Disable current line highlighting
      hideCursorInOverviewRuler: true,  // Remove cursor indictor in right hand side scroll bar
      readOnly: qwebrOptions['read-only'] ?? false
    });

    // Store the official counter ID to be used in keyboard shortcuts
    editor.__qwebrCounter = qwebrCounter;

    // Store the official div container ID
    editor.__qwebrEditorId = `qwebr-editor-${qwebrCounter}`;

    // Store the initial code value and options
    editor.__qwebrinitialCode = initialCode;
    editor.__qwebrOptions = qwebrOptions;

    // Set at the model level the preferred end of line (EOL) character to LF.
    // This prevent `\r\n` from being given to the webR engine if the user is on Windows.
    // See details in: https://github.com/coatless/quarto-webr/issues/94
    // Associated error text: 
    // Error: <text>:1:7 unexpected input

    // Retrieve the underlying model
    const model = editor.getModel();
    // Set EOL for the model
    model.setEOL(monaco.editor.EndOfLineSequence.LF);

    // Dynamically modify the height of the editor window if new lines are added.
    let ignoreEvent = false;
    const updateHeight = () => {
      const contentHeight = editor.getContentHeight();
      // We're avoiding a width change
      //editorDiv.style.width = `${width}px`;
      editorDiv.style.height = `${contentHeight}px`;
      try {
        ignoreEvent = true;

        // The key to resizing is this call
        editor.layout();
      } finally {
        ignoreEvent = false;
      }
    };

    // Helper function to check if selected text is empty
    function isEmptyCodeText(selectedCodeText) {
      return (selectedCodeText === null || selectedCodeText === undefined || selectedCodeText === "");
    }

    // Registry of keyboard shortcuts that should be re-added to each editor window
    // when focus changes.
    const addWebRKeyboardShortCutCommands = () => {
      // Add a keydown event listener for Shift+Enter to run all code in cell
      editor.addCommand(monaco.KeyMod.Shift | monaco.KeyCode.Enter, () => {

        // Retrieve all text inside the editor
        qwebrExecuteCode(editor.getValue(), editor.__qwebrCounter, editor.__qwebrOptions);
      });

      // Add a keydown event listener for CMD/Ctrl+Enter to run selected code
      editor.addCommand(monaco.KeyMod.CtrlCmd | monaco.KeyCode.Enter, () => {

        // Get the selected text from the editor
        const selectedText = editor.getModel().getValueInRange(editor.getSelection());
        // Check if no code is selected
        if (isEmptyCodeText(selectedText)) {
          // Obtain the current cursor position
          let currentPosition = editor.getPosition();
          // Retrieve the current line content
          let currentLine = editor.getModel().getLineContent(currentPosition.lineNumber);

          // Propose a new position to move the cursor to
          let newPosition = new monaco.Position(currentPosition.lineNumber + 1, 1);

          // Check if the new position is beyond the last line of the editor
          if (newPosition.lineNumber > editor.getModel().getLineCount()) {
            // Add a new line at the end of the editor
            editor.executeEdits("addNewLine", [{
            range: new monaco.Range(newPosition.lineNumber, 1, newPosition.lineNumber, 1),
            text: "\n", 
            forceMoveMarkers: true,
            }]);
          }
          
          // Run the entire line of code.
          qwebrExecuteCode(currentLine, editor.__qwebrCounter, editor.__qwebrOptions);

          // Move cursor to new position
          editor.setPosition(newPosition);
        } else {
          // Code to run when Ctrl+Enter is pressed with selected code
          qwebrExecuteCode(selectedText, editor.__qwebrCounter, editor.__qwebrOptions);
        }
      });
    }

    // Register an on focus event handler for when a code cell is selected to update
    // what keyboard shortcut commands should work.
    // This is a workaround to fix a regression that happened with multiple
    // editor windows since Monaco 0.32.0 
    // https://github.com/microsoft/monaco-editor/issues/2947
    editor.onDidFocusEditorText(addWebRKeyboardShortCutCommands);

    // Register an on change event for when new code is added to the editor window
    editor.onDidContentSizeChange(updateHeight);

    // Manually re-update height to account for the content we inserted into the call
    updateHeight();

    // Store the editor instance in the global dictionary
    qwebrEditorInstances[editor.__qwebrCounter] = editor;

  });

  // Add a click event listener to the run button
  runButton.onclick = function () {
    qwebrExecuteCode(editor.getValue(), editor.__qwebrCounter, editor.__qwebrOptions);
  };

  // Add a click event listener to the reset button
  copyButton.onclick = function () {
    // Retrieve current code data
    const data = editor.getValue();
    
    // Write code data onto the clipboard.
    navigator.clipboard.writeText(data || "");
  };
  
  // Add a click event listener to the copy button
  resetButton.onclick = function () {
    editor.setValue(editor.__qwebrinitialCode);
  };
  
}
</script>
  <div class="reveal">
    <div class="slides">


<section>
<section id="an-exploration-of-information-loss-in-transformer-embedding-spaces-for-enhancing-predictive-ai-in-genomics" class="title-slide slide level1 center" data-background-image="images/DALL_E_DNA_Image_edited.png" data-background-size="cover" data-background-color="#4f6952">
<h1><span style="color:#FFFFFF;           font-size: 55px;           position: relative;          bottom: 60px;           text-shadow: -1px -1px 0 #000,                        1px -1px 0 #000,                        -1px 1px 0 #000,                        1px 1px 0 #000;">An Exploration of Information Loss in Transformer Embedding Spaces for Enhancing Predictive AI in Genomics</span></h1>
<h1 style="color:#FFFFFF; 
         font-size: 28px; 
         text-align: center;
         position: absolute; 
         top: 325px; 
         width: 100%; 
         text-shadow: -0.75px -0.75px 0 #000, 
                      0.75px -0.75px 0 #000, 
                      -0.75px 0.75px 0 #000, 
                      0.75px 0.75px 0 #000;">
Daniel Hintz <br> 2024-06-06
</h1>
</section>
<section id="thesis-question" class="title-slide slide level2 center">
<h2>Thesis Question</h2>
<div class="fragment">
<div style="text-align:center; font-size: 0.85em;">
<p>How easily can information be extracted from GenSLM embeddings while maintaining its original integrity; and what is the quality of the produced vector embeddings?</p>
</div>
<aside class="notes">
<ul>
<li>The jargon won’t have meaning now, will very soon, i.e.
<ul>
<li>What is GenSLM</li>
<li>What is an embedding</li>
<li>Why do we care about the quality of an embedding</li>
<li>What are real world applications<br>
</li>
</ul></li>
</ul>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</div>
</section>
</section>
<section id="outline" class="title-slide slide level1 center">
<h1><span style="font-size: 55px;">Outline</span></h1>
<div style="font-size: 0.65em;">
<ul>
<li>Thesis Question</li>
<li>Background
<ul>
<li>DNA</li>
<li>Motivation</li>
<li>GenSLM Embedding Algorithm</li>
</ul></li>
<li>Data and Processing
<ul>
<li>Source and Cleaning</li>
</ul></li>
<li>Methodology
<ul>
<li>Intrinsic and Extrinsic Evaluation</li>
</ul></li>
<li>Results</li>
<li>Conclusion</li>
</ul>
</div>
</section>

<section>
<section id="background" class="title-slide slide level1 center" data-background-color="#8FA9C2">
<h1><span style="color:white;">Background</span></h1>

</section>
<section id="dna" class="title-slide slide level2 center">
<h2>DNA</h2>
<div class="columns">
<div class="column" style="width:80%;">
<div>
<div style="font-size: 0.6em;">
<ul>
<li class="fragment">Deoxyribonucleic acid (DNA) is a molecule that contains the genetic code that provides the instructions for building and maintaining life.</li>
<li class="fragment">The structure of DNA can be thought of as rungs on a ladder (known as base pairs) involving the pairing of four nucleotides - Adenine (A), Cytosine (C),Guanine (G) and Thymine (T).</li>
</ul>
</div>
</div>
</div><div class="column" style="width:20%;">
<div style="text-align:center; font-size: 0.5em;">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="images/DNA_diagram.png"></p>
<figcaption>Figure 1: DNA Base Pairs; <span class="citation" data-cites="NHGRI2024">[<a href="#/references" role="doc-biblioref" onclick="">1</a>]</span></figcaption>
</figure>
</div>
</div>
</div>
</div>
</section>

<section id="dna-sequencing-technology" class="title-slide slide level2 center">
<h2>DNA Sequencing Technology</h2>
<div class="columns">
<div class="column" style="width:50%;">
<div>
<div style="font-size: 0.6em;">
<ul>
<li class="fragment">Someone gets tested for Covid, PCA is ran to detect if the assay is indeed positive.</li>
<li class="fragment">Positive tests are sequenced using a machine that is most likely either Ilumina or Oxford Nanopore.</li>
<li class="fragment">This takes in the Covid Sample and arrives at a digital copy of DNA sequences.</li>
</ul>
</div>
</div>
</div><div class="column" style="width:50%;">
<div style="text-align:center; font-size: 0.5em;">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="images/DNA_sequencing.png"></p>
<figcaption>Figure 2: Simplified Schema of Sequencing Process for SARS-CoV-2</figcaption>
</figure>
</div>
</div>
</div>
</div>
</section>

<section id="sars-cov-2-and-proteins" class="title-slide slide level2 center">
<h2>SARS-CoV-2 and Proteins</h2>
<div class="columns">
<div class="column" style="width:40%;">
<div>
<div style="font-size: 0.6em;">
<ul>
<li class="fragment">SARS-CoV-2 has 29 proteins.</li>
<li class="fragment">Different proteins have different functions.</li>
<li class="fragment">Proteins are encoded from different sites of DNA.</li>
<li class="fragment">In studying embeddings, tasks have been either <strong>whole genome oriented</strong> or <strong>protein specific</strong>. <!--
- In the context of SARS-CoV-2, proteins are responsible for the virus's structure and infectivity.
-  Nonstructural Proteins primarily function in the replication and manipulation of the host cell's environment to favor viral replication.
--></li>
</ul>
</div>
</div>
</div><div class="column" style="width:60%;">
<div style="text-align:center; font-size: 0.5em;">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="images/nsp1_str_non_str.png"></p>
<figcaption>Figure 3: SARS-CoV-2 Proteins; Adapted from <span class="citation" data-cites="kandwal2023genetic">[<a href="#/references" role="doc-biblioref" onclick="">2</a>]</span>, pg. 99</figcaption>
</figure>
</div>
</div>
</div>
</div>
</section>

<section id="mutations" class="title-slide slide level2 center">
<h2>Mutations</h2>
<div class="columns">
<div class="column" style="width:40%;">
<div>
<div style="font-size: 0.6em;">
<ul>
<li class="fragment">A mutation in the context of DNA refers to a change in the nucleotide sequence of the genetic material of an organism.</li>
<li class="fragment">Mutations can be random or induced by external factors.</li>
<li class="fragment">For contect, SARS-CoV-2 can mutate to become more infective for increasing its chances of survival.</li>
<li class="fragment">There many different types of mutations but only substitutions are illustrated for this talk.</li>
</ul>
</div>
</div>
</div><div class="column" style="width:60%;">
<div style="text-align:center; font-size: 0.5em;">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="images/mutation_example.png" style="width:45.0%"></p>
<figcaption>Figure 4: Substitution Mutations</figcaption>
</figure>
</div>
</div>
</div>
</div>
<aside class="notes">
<ul>
<li>Spontaneous mutations: These arise naturally and randomly without any external influence, often due to errors in DNA replication. Errors might include mispairing of nucleotides or slippage during replication.</li>
<li>Induced mutations: These are caused by external factors, known as mutagens, which can include physical agents like ultraviolet light and X-rays, or chemical agents such as certain drugs and pollutants.</li>
</ul>
<p><strong>Why Mutations Occur</strong></p>
<p>Random errors: Many mutations result simply from random errors during DNA replication. DNA polymerase, the enzyme responsible for replicating DNA, is not perfect and occasionally makes mistakes.</p>
<p>Environmental stress: Exposure to environmental stressors like radiation, chemicals, and even biological agents can damage DNA and lead to mutations.</p>
<p>Evolutionary advantage: From an evolutionary perspective, mutations are a key mechanism of genetic variability and thus, evolution. While many mutations are neutral or harmful, some can confer advantages that lead to positive selection in a population.</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>

<section id="motivation" class="title-slide slide level2 center">
<h2>Motivation</h2>
<div>
<div style="font-size: 0.6em;">
<ul>
<li class="fragment">This project came about from connecting with Nick Chia, a biophysicist at Argonne National labs.</li>
<li class="fragment">Nick and his collaborators in 2020 (then at the Mayo Clinic) made an algorithm that predicted the trajectory of colorectal cancer mutations.</li>
<li class="fragment">However, this algorithm was slow and expensive to train.</li>
<li class="fragment"><strong>The embeddings used in this algorithm was one of the biggest choke points</strong>.</li>
<li class="fragment">New and improved ways are needed to embed large genomes into lower dimensions
<ul>
<li class="fragment">For example, a one hot encoding for the human genome as DNA dimensionally is n x 3 billion x 4, where n is the number of patients, 3 billion is the number of rows for each nucleotide and 4 columns represent the four possible bases (A, T, G, C).</li>
</ul></li>
<li class="fragment">This is the <strong>Big P little N problem</strong>; hence dimension reduction via an embedding is extremely advantageous!</li>
<li class="fragment">Further work needs to be done is studying the quality of embeddings to help bring about more efficient real world implementations like precision medicine for colorectal cancer.</li>
</ul>
</div>
</div>
</section>

<section id="dimension-reduction" class="title-slide slide level2 center">
<h2>Dimension Reduction</h2>
<div style="font-size: 0.6em;">
<p>In the UW Statistics program what methods do we learn to apply dimension reduction?</p>
</div>
<div>
<div style="font-size: 0.6em;">
<ul>
<li class="fragment">Principal component analysis (PCA)!</li>
<li class="fragment">PCA as well as embeddings, transform the coordinate system of our data.</li>
<li class="fragment">The difference for this project is that dimension reduction is being applied to DNA.</li>
</ul>
</div>
</div>
</section>

<section id="embeddings" class="title-slide slide level2 center">
<h2>Embeddings</h2>
<div>
<div style="font-size: 0.55em;">
<ul>
<li class="fragment"><p>An <em>embedding</em> is the name for the dense numeric vectors produced by <em>embedding algorithms</em>, such as GenSLM.</p></li>
<li class="fragment"><p>“Embedding” generally refers to the embedding vector.</p></li>
<li class="fragment"><p>Embeddings are representation of data in lower-dimensional space that preserves some aspects of the original structure of the data.</p></li>
<li class="fragment"><p>But why would you ever embed something?</p>
<ul>
<li class="fragment">Embedding vectors are generally more computationally efficient.</li>
</ul></li>
</ul>
</div>
</div>
<aside class="notes">
<ul>
<li>As an introduction, it will be easier to first walk through an example of an Embedding for natural language.</li>
</ul>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>

<section id="sequence-embeddings" class="title-slide slide level2 center">
<h2>Sequence Embeddings</h2>
<div>
<div style="font-size: 0.6em;">
<ul>
<li class="fragment">Neural network embedding algorithms for genomic data include <strong>GenSLM (the focus of this presentation)</strong>, DNABERT2, and HyenaDNA.</li>
<li class="fragment">These embeddings are lossy encodings, meaning that some information is lost in the transformation.</li>
<li class="fragment">They also can distort structural relationships represented in the data.</li>
</ul>
</div>
</div>
<aside class="notes">
<ul>
<li>Just like we can represent words as vectors we can also represent genomic sequences as vectors.</li>
</ul>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
<!--
## Why Data Representation Matters 

::: incremental
::: {style="font-size: 0.6em;"}
- An embedding is one way of representing data.
:::
:::

. . .

::: {style="text-align:center; font-size: 0.5em;"}
![](images/pred_rep_obj.png){width="65%"}
:::

::: {style="margin-top: 20px;"}
:::

::: incremental
::: {style="font-size: 0.6em;"}
-  An effective representation (embedding) of a genomic sequence is crucial for all downstream tasks.
    - i.e., clustering, classification, regression, protein function identification, structural analysis, and predicting genetic disorders [@angermueller2016deep].
- Given the increasing use of embeddings, embedding evaluation has also become more important. 
:::
:::
-->
</section>

<section id="genslm" class="title-slide slide level2 center">
<h2>GenSLM</h2>
<div class="panel-tabset">
<ul id="tabset-1" class="panel-tabset-tabby"><li><a data-tabby-default="" href="#tabset-1-1">Overview</a></li><li><a href="#tabset-1-2">Transformer</a></li><li><a href="#tabset-1-3">Diffusion + Transformer</a></li></ul>
<div class="tab-content">
<div id="tabset-1-1">
<div class="columns">
<div class="column" style="width:40%;">
<div>
<div style="font-size: 0.6em;">
<ul>
<li class="fragment">Overall, the GenSLM algorithm first <strong>tokenizes</strong>, i.e., breaks up sequences in to chunks of three nucleotides.</li>
<li class="fragment">Then, the input sequence are <strong>vectorized</strong> creating a <span class="math inline">\(1 \times 512\)</span> vector for each inputed sequence.</li>
</ul>
</div>
</div>
</div><div class="column" style="width:60%;">
<div style="text-align:center; font-size: 0.5em;">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="images/Tokenize_vectorize.png" style="width:50.0%"></p>
<figcaption>Figure 5: Generic Embedding Workflow</figcaption>
</figure>
</div>
</div>
</div>
</div>
</div>
<div id="tabset-1-2">
<div class="columns">
<div class="column" style="width:40%;">
<div>
<div style="font-size: 0.6em;">
<ul>
<li class="fragment">First the tokenized input sequence is passed to the transformer encoder.</li>
<li class="fragment">The transformer encoder converts 1,024 bp slices into numeric vectors.</li>
<li class="fragment">This is ran recursively through a diffusion model to learn a condensed distribution of the whole sequence.</li>
<li class="fragment">The transformers in GenSLM are used to capture local interactions within a genomic sequence.</li>
</ul>
</div>
</div>
</div><div class="column" style="width:60%;">
<div style="text-align:center; font-size: 0.5em;">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="images/GenSLM.png"></p>
<figcaption>Figure 6: GenSLM Transformer Architecture; <span class="citation" data-cites="genslms2023">[<a href="#/references" role="doc-biblioref" onclick="">3</a>]</span></figcaption>
</figure>
</div>
</div>
</div>
</div>
</div>
<div id="tabset-1-3">
<div>
<div style="font-size: 0.6em;">
<ul>
<li class="fragment">Diffusion and Transformer models work in tandem in GenSLM.</li>
<li class="fragment">Transformers captures local interactions, while the diffusion model captures learns a representation of transformers vectors to represent the whole genome.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
</section>
</section>
<section>
<section id="downstream" class="title-slide slide level1 center" data-background-color="#8FA9C2">
<h1><span style="color:white;">Downstream</span></h1>

</section>
<section id="downstream-1" class="title-slide slide level2 center">
<h2>Downstream</h2>
<div>
<div style="font-size: 0.6em;">
<ul>
<li class="fragment"><p>Consider our colereactal cancer example, in the context of precision medicine, you may want to classify aggressive and non-aggressive cancers.</p></li>
<li class="fragment"><p>Or in the context of SARS-CoV-2, you may want to be able to monitor changes in variants and how they differ from previous variants.</p></li>
<li class="fragment"><p>Assuming embeddings in both examples, the tasks are considered <strong>downstream</strong> from the process of creating the embeddings from the original data.</p></li>
</ul>
</div>
</div>
</section>

<section id="raw-dna-versus-embedded-data" class="title-slide slide level2 center">
<h2>Raw DNA versus embedded Data</h2>
<div id="qwebr-insertion-location-1"></div>
<noscript>Please enable JavaScript to experience the dynamic code cell content on this page.</noscript>
<!--
# DT::datatable(Whole_genome_covid_seqs)
-->
<div class="cell">
<div class="cell-output-display">

<div class="datatables html-widget html-fill-item" id="htmlwidget-fc9496453d837b7ba89e" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-fc9496453d837b7ba89e">{"x":{"filter":"none","vertical":false,"data":[["1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18","19","20","21","22","23","24","25","26","27","28","29","30","31","32","33","34","35","36","37","38","39","40","41","42","43","44","45","46","47","48","49","50","51","52","53","54","55","56","57","58","59","60","61","62","63","64","65","66","67","68","69","70","71","72","73","74","75","76","77","78","79","80","81","82","83","84","85","86","87","88","89","90","91","92","93","94","95","96","97","98","99","100","101","102","103","104","105","106","107","108","109","110","111","112","113","114","115","116","117","118","119","120","121","122","123","124","125","126","127","128","129","130","131","132","133","134","135","136","137","138","139","140","141","142","143","144","145","146","147","148","149","150"],[5.1,4.9,4.7,4.6,5,5.4,4.6,5,4.4,4.9,5.4,4.8,4.8,4.3,5.8,5.7,5.4,5.1,5.7,5.1,5.4,5.1,4.6,5.1,4.8,5,5,5.2,5.2,4.7,4.8,5.4,5.2,5.5,4.9,5,5.5,4.9,4.4,5.1,5,4.5,4.4,5,5.1,4.8,5.1,4.6,5.3,5,7,6.4,6.9,5.5,6.5,5.7,6.3,4.9,6.6,5.2,5,5.9,6,6.1,5.6,6.7,5.6,5.8,6.2,5.6,5.9,6.1,6.3,6.1,6.4,6.6,6.8,6.7,6,5.7,5.5,5.5,5.8,6,5.4,6,6.7,6.3,5.6,5.5,5.5,6.1,5.8,5,5.6,5.7,5.7,6.2,5.1,5.7,6.3,5.8,7.1,6.3,6.5,7.6,4.9,7.3,6.7,7.2,6.5,6.4,6.8,5.7,5.8,6.4,6.5,7.7,7.7,6,6.9,5.6,7.7,6.3,6.7,7.2,6.2,6.1,6.4,7.2,7.4,7.9,6.4,6.3,6.1,7.7,6.3,6.4,6,6.9,6.7,6.9,5.8,6.8,6.7,6.7,6.3,6.5,6.2,5.9],[3.5,3,3.2,3.1,3.6,3.9,3.4,3.4,2.9,3.1,3.7,3.4,3,3,4,4.4,3.9,3.5,3.8,3.8,3.4,3.7,3.6,3.3,3.4,3,3.4,3.5,3.4,3.2,3.1,3.4,4.1,4.2,3.1,3.2,3.5,3.6,3,3.4,3.5,2.3,3.2,3.5,3.8,3,3.8,3.2,3.7,3.3,3.2,3.2,3.1,2.3,2.8,2.8,3.3,2.4,2.9,2.7,2,3,2.2,2.9,2.9,3.1,3,2.7,2.2,2.5,3.2,2.8,2.5,2.8,2.9,3,2.8,3,2.9,2.6,2.4,2.4,2.7,2.7,3,3.4,3.1,2.3,3,2.5,2.6,3,2.6,2.3,2.7,3,2.9,2.9,2.5,2.8,3.3,2.7,3,2.9,3,3,2.5,2.9,2.5,3.6,3.2,2.7,3,2.5,2.8,3.2,3,3.8,2.6,2.2,3.2,2.8,2.8,2.7,3.3,3.2,2.8,3,2.8,3,2.8,3.8,2.8,2.8,2.6,3,3.4,3.1,3,3.1,3.1,3.1,2.7,3.2,3.3,3,2.5,3,3.4,3],[1.4,1.4,1.3,1.5,1.4,1.7,1.4,1.5,1.4,1.5,1.5,1.6,1.4,1.1,1.2,1.5,1.3,1.4,1.7,1.5,1.7,1.5,1,1.7,1.9,1.6,1.6,1.5,1.4,1.6,1.6,1.5,1.5,1.4,1.5,1.2,1.3,1.4,1.3,1.5,1.3,1.3,1.3,1.6,1.9,1.4,1.6,1.4,1.5,1.4,4.7,4.5,4.9,4,4.6,4.5,4.7,3.3,4.6,3.9,3.5,4.2,4,4.7,3.6,4.4,4.5,4.1,4.5,3.9,4.8,4,4.9,4.7,4.3,4.4,4.8,5,4.5,3.5,3.8,3.7,3.9,5.1,4.5,4.5,4.7,4.4,4.1,4,4.4,4.6,4,3.3,4.2,4.2,4.2,4.3,3,4.1,6,5.1,5.9,5.6,5.8,6.6,4.5,6.3,5.8,6.1,5.1,5.3,5.5,5,5.1,5.3,5.5,6.7,6.9,5,5.7,4.9,6.7,4.9,5.7,6,4.8,4.9,5.6,5.8,6.1,6.4,5.6,5.1,5.6,6.1,5.6,5.5,4.8,5.4,5.6,5.1,5.1,5.9,5.7,5.2,5,5.2,5.4,5.1],[0.2,0.2,0.2,0.2,0.2,0.4,0.3,0.2,0.2,0.1,0.2,0.2,0.1,0.1,0.2,0.4,0.4,0.3,0.3,0.3,0.2,0.4,0.2,0.5,0.2,0.2,0.4,0.2,0.2,0.2,0.2,0.4,0.1,0.2,0.2,0.2,0.2,0.1,0.2,0.2,0.3,0.3,0.2,0.6,0.4,0.3,0.2,0.2,0.2,0.2,1.4,1.5,1.5,1.3,1.5,1.3,1.6,1,1.3,1.4,1,1.5,1,1.4,1.3,1.4,1.5,1,1.5,1.1,1.8,1.3,1.5,1.2,1.3,1.4,1.4,1.7,1.5,1,1.1,1,1.2,1.6,1.5,1.6,1.5,1.3,1.3,1.3,1.2,1.4,1.2,1,1.3,1.2,1.3,1.3,1.1,1.3,2.5,1.9,2.1,1.8,2.2,2.1,1.7,1.8,1.8,2.5,2,1.9,2.1,2,2.4,2.3,1.8,2.2,2.3,1.5,2.3,2,2,1.8,2.1,1.8,1.8,1.8,2.1,1.6,1.9,2,2.2,1.5,1.4,2.3,2.4,1.8,1.8,2.1,2.4,2.3,1.9,2.3,2.5,2.3,1.9,2,2.3,1.8],["setosa","setosa","setosa","setosa","setosa","setosa","setosa","setosa","setosa","setosa","setosa","setosa","setosa","setosa","setosa","setosa","setosa","setosa","setosa","setosa","setosa","setosa","setosa","setosa","setosa","setosa","setosa","setosa","setosa","setosa","setosa","setosa","setosa","setosa","setosa","setosa","setosa","setosa","setosa","setosa","setosa","setosa","setosa","setosa","setosa","setosa","setosa","setosa","setosa","setosa","versicolor","versicolor","versicolor","versicolor","versicolor","versicolor","versicolor","versicolor","versicolor","versicolor","versicolor","versicolor","versicolor","versicolor","versicolor","versicolor","versicolor","versicolor","versicolor","versicolor","versicolor","versicolor","versicolor","versicolor","versicolor","versicolor","versicolor","versicolor","versicolor","versicolor","versicolor","versicolor","versicolor","versicolor","versicolor","versicolor","versicolor","versicolor","versicolor","versicolor","versicolor","versicolor","versicolor","versicolor","versicolor","versicolor","versicolor","versicolor","versicolor","versicolor","virginica","virginica","virginica","virginica","virginica","virginica","virginica","virginica","virginica","virginica","virginica","virginica","virginica","virginica","virginica","virginica","virginica","virginica","virginica","virginica","virginica","virginica","virginica","virginica","virginica","virginica","virginica","virginica","virginica","virginica","virginica","virginica","virginica","virginica","virginica","virginica","virginica","virginica","virginica","virginica","virginica","virginica","virginica","virginica","virginica","virginica","virginica","virginica","virginica","virginica"]],"container":"<table class=\"cell-border stripe\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>Sepal.Length<\/th>\n      <th>Sepal.Width<\/th>\n      <th>Petal.Length<\/th>\n      <th>Petal.Width<\/th>\n      <th>Species<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"pageLength":4,"dom":"tip","columnDefs":[{"className":"dt-right","targets":[1,2,3,4]},{"orderable":false,"targets":0},{"name":" ","targets":0},{"name":"Sepal.Length","targets":1},{"name":"Sepal.Width","targets":2},{"name":"Petal.Length","targets":3},{"name":"Petal.Width","targets":4},{"name":"Species","targets":5}],"order":[],"autoWidth":false,"orderClasses":false,"lengthMenu":[4,10,25,50,100],"rowCallback":"function(row, data, displayNum, displayIndex, dataIndex) {\nvar value=data[1]; $(this.api().cell(row, 1).node()).css({'font-size':'10px'});\nvar value=data[2]; $(this.api().cell(row, 2).node()).css({'font-size':'10px'});\nvar value=data[3]; $(this.api().cell(row, 3).node()).css({'font-size':'10px'});\nvar value=data[4]; $(this.api().cell(row, 4).node()).css({'font-size':'10px'});\nvar value=data[5]; $(this.api().cell(row, 5).node()).css({'font-size':'10px'});\n}"},"callback":"function(table) {\ntable.columns().header().to$().css({'font-size': '10px'});\n}"},"evals":["options.rowCallback","callback"],"jsHooks":[]}</script>

</div>
</div>
</section>

<section id="project-workflow" class="title-slide slide level2 center">
<h2>Project Workflow</h2>
<div style="text-align:center; font-size: 0.5em;">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="images/Overall_workflow.png" style="width:30.0%"></p>
<figcaption>Figure 7: Project Workflow</figcaption>
</figure>
</div>
</div>
</section>
</section>
<section>
<section id="data-and-processing" class="title-slide slide level1 center" data-background-color="#8FA9C2">
<h1><span style="color:white;">Data and Processing</span></h1>

</section>
<section id="data-description" class="title-slide slide level2 center">
<h2>Data Description</h2>
<div>
<div style="font-size: 0.6em;">
<ul>
<li class="fragment">All Covid sequences were downloaded with permission from the Global Initiative on Sharing All Influenza Data (GISAID).</li>
<li class="fragment">No geographical or temporal restrictions were placed on the sequences extracted.</li>
</ul>
</div>
</div>
</section>

<section id="data-cleaning" class="title-slide slide level2 center">
<h2>Data Cleaning</h2>
<div>
<div style="font-size: 0.6em;">
<ul>
<li class="fragment">GISAID’s exclusion criteria was used to remove sequences of poor quality.</li>
<li class="fragment">Additional exclusion criteria was implemented to remove all sequence with ambiguous nucleotides (the <code>NA's</code> of the genomic world) and large gap sizes post-alignment.</li>
</ul>
</div>
</div>
</section>

<section id="multiple-sequence-alignment" class="title-slide slide level2 center">
<h2>Multiple Sequence Alignment</h2>
<div>
<div style="font-size: 0.6em;">
<ul>
<li class="fragment">An multiple alignment alignment was performed to be able to slice the exact location of proteins across different sequences.</li>
</ul>
</div>
</div>
<div class="fragment">
<div style="text-align:center; font-size: 0.5em;">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="images/Unaligned_example.png" style="width:30.0%"></p>
<figcaption>Figure 18: Un-aligned Sequences</figcaption>
</figure>
</div>
</div>
</div>
<div class="fragment">
<div style="text-align:center; font-size: 0.5em;">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="images/aligned_example.png" style="width:30.0%"></p>
<figcaption>Figure 19: Aligned Sequences</figcaption>
</figure>
</div>
</div>
</div>
</section>
</section>
<section>
<section id="exploratory-data-analysis" class="title-slide slide level1 center" data-background-color="#8FA9C2">
<h1><span style="color:white;">Exploratory Data Analysis</span></h1>
<!--
## [What are the Patterns and Structures Oberved from the Sequence Data?]{style="font-size: 52px;"}

::: incremental
::: {style="font-size: 0.6em;"}
- For Exploratory Data Analysis, the goal was the characterize the structure of the data in its sequence representation. 
-  This laid the groundwork for subsequent comparisons with the embedding structure.
:::
::: 

. . .

::: {style="text-align:center; font-size: 0.5em;"}
![Figure 8: Mutation Boxplot](images/Whole_Genome_Mutation_Count_Hamming_Distances.png){width="50%"}
:::
-->
</section>
<section id="section" class="title-slide slide level2 center">
<h2></h2>
<div style="text-align:center; font-size: 0.5em;">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="images/hamming_prot_heatmap_581_all_variants.png" style="width:100.0%"></p>
<figcaption>Figure 11: Hamming Distance of Aligned Proteins to Wuhan Reference Sequence</figcaption>
</figure>
</div>
</div>
</section>
</section>
<section>
<section id="methodology" class="title-slide slide level1 center" data-background-color="#8FA9C2">
<h1><span style="color:white;">Methodology</span></h1>

</section>
<section id="evaluating-embeddings" class="title-slide slide level2 center">
<h2>Evaluating Embeddings</h2>
<div>
<div style="font-size: 0.6em;">
<ul>
<li class="fragment"><p>Broadly speaking, the quality of an embedding is assessed based upon its information richness and the degree of non-redundancy.</p></li>
<li class="fragment"><p>Their are two main methodologies for evaluating embeddings:</p>
<ul>
<li class="fragment">When downstream learning tasks are evaluated, it’s referred to as <strong>extrinsic evaluation</strong> <span class="citation" data-cites="lavravc2021representation">[<a href="#/references" role="doc-biblioref" onclick="">4</a>]</span>.</li>
<li class="fragment">When the qualities of the embedding matrix itself are assessed it is referred to as <strong>intrinsic evaluation</strong> <span class="citation" data-cites="lavravc2021representation">[<a href="#/references" role="doc-biblioref" onclick="">4</a>]</span>.</li>
</ul></li>
</ul>
</div>
</div>
</section>

<section id="intrinsic-evaluation" class="title-slide slide level2 center">
<h2>Intrinsic Evaluation</h2>
<div>
<div style="font-size: 0.6em;">
<ul>
<li class="fragment">There are three main qualities to asses when studying the the quality of an embedding: <strong>Redundancy</strong>, <strong>Separability</strong>, <strong>Preservation of Semantic Distance</strong>
<ul>
<li class="fragment">Redundancy reveals the efficiency of a embeddings encoding; more efficient encodings tend to perform better and use fewer computational resources.</li>
<li class="fragment">Separability gives practical insight into whether or not the embedding output can be separated into meaningful genomic groups (i.e., variants).</li>
<li class="fragment">Preservation of semantic distance is important for determining if the structural representation of information remains valid for subsequent tasks.</li>
</ul></li>
<li class="fragment">Sub-methods used for intrinsic evaluation include Singular Value Decomposition (SVD), Distance matrices, Radial Dendograms, and Principal Copnonent Analysis.</li>
</ul>
</div>
</div>
</section>

<section id="extrinsic-evaluation" class="title-slide slide level2 center">
<h2>Extrinsic Evaluation</h2>
<div>
<div style="font-size: 0.6em;">
<ul>
<li class="fragment">Extrinsic evaluation pursues practical benchmarks to assess the performance of an embeddings algorithm per its associated embedding matrix across varying levels of task complexity.</li>
<li class="fragment">For GenSLM, the subset of possible tasks chosen were variant and protein classification.</li>
<li class="fragment">A Classification and Regression Tree (CART) was used for performing the classification.</li>
</ul>
</div>
</div>
</section>
</section>
<section>
<section id="results" class="title-slide slide level1 center" data-background-color="#8FA9C2">
<h1><span style="color:white;">Results</span></h1>

</section>
<section id="intrinsic-evaluation-results" class="title-slide slide level2 center">
<h2><span style="color:#6FB1D1;">Intrinsic Evaluation Results</span></h2>

</section>
<section id="redundancy" class="title-slide slide level3 center">
<h3><span style="color:#A9D0E3;">Redundancy</span></h3>

</section>
<section id="singular-value-decomposition" class="slide level4">
<h4><span style="font-size: 0.8em;">Singular Value Decomposition</span></h4>
<div style="text-align:center; font-size: 0.5em;">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="images/SVD_582_cpe_2.png" style="width:92.0%"></p>
<figcaption>Figure 13: SVD CPE Plot</figcaption>
</figure>
</div>
</div>
</section>

<section id="preservation-of-semantic-distance" class="title-slide slide level3 center">
<h3><span style="color:#A9D0E3;">Preservation of Semantic Distance</span></h3>

</section>
<section id="distance-matrices" class="slide level4">
<h4>Distance Matrices</h4>
<!--X axis labels-->
<p><img data-src="images/variant_labels/alpha.png" class="absolute" style="top: 675px; left: 110px; width: 20.55px; height: 16.91px; "> <img data-src="images/variant_labels/beta.png" class="absolute" style="top: 675px; left: 228px; width: 18.17px; height: 29.48px; "> <img data-src="images/variant_labels/delta.png" class="absolute" style="top: 675px; left: 340px; width: 14.64px; height: 24.91px; "> <img data-src="images/variant_labels/epsilon.png" class="absolute" style="top: 675px; left: 455px; width: 15.74px; height: 21.83px; "> <img data-src="images/variant_labels/gamma.png" class="absolute" style="top: 675px; left: 575px; width: 20.07px; height: 25.14px; "></p>
<p><img data-src="images/variant_labels/mu.png" class="absolute" style="top: 675px; left: 692px; width: 20.71px; height: 25.14px; "> <img data-src="images/variant_labels/omicron.png" class="absolute" style="top: 675px; left: 815px; width: 21.45px; height: 22.91px; "></p>
<!--Y axis labels-->
<p><img data-src="images/variant_labels/alpha_90degrees.png" class="absolute" style="top: 630px; left: 37.5px; width: 16.91px; height: 20.55px; "> <img data-src="images/variant_labels/beta_90degrees.png" class="absolute" style="top: 580px; left: 32px; width: 29.48px; height: 18.17px; "></p>
<p><img data-src="images/variant_labels/delta_90degrees.png" class="absolute" style="top: 538px; left: 32px; width: 24.91px; height: 14.64px; "></p>
<p><img data-src="images/variant_labels/epsilon_90degrees.png" class="absolute" style="top: 490px; left: 32px; width: 21.83px; height: 15.74px; "></p>
<p><img data-src="images/variant_labels/gamma_90degrees.png" class="absolute" style="top: 440px; left: 32px; width: 25.14px; height: 20.07px; "></p>
<p><img data-src="images/variant_labels/mu_90degrees.png" class="absolute" style="top: 390px; left: 32px; width: 25.14px; height: 20.71px; "></p>
<p><img data-src="images/variant_labels/omicron_90degrees.png" class="absolute" style="top: 340px; left: 32px; width: 22.91px; height: 21.45px; "></p>
<div class="panel-tabset">
<ul id="tabset-2" class="panel-tabset-tabby"><li><a data-tabby-default="" href="#tabset-2-1">Sequence</a></li><li><a href="#tabset-2-2">Embeddings</a></li><li><a href="#tabset-2-3">Abs. Diff.</a></li><li><a href="#tabset-2-4">Reg. Diff.</a></li></ul>
<div class="tab-content">
<div id="tabset-2-1">
<div style="margin-top: 0px;">

</div>
<div style="font-size: 0.6em;">
<ul>
<li>There is a lot of heterogeneity in the differences among sequences as fine scales.</li>
<li>Notice the L shape in striations in alpha sequences compare to other variants.</li>
</ul>
</div>
<iframe class="plotlyFrame" src="./seq_dist_matrix.html" style="border: none;">
</iframe>
</div>
<div id="tabset-2-2">
<div style="margin-top: 5.5px;">

</div>
<div style="font-size: 0.6em;">
<ul>
<li>Comparing the Sequence distance matrix to the Embedding distance matrix we see that finer differences amongst sequences are lost in the embedding transformation.</li>
</ul>
</div>
<iframe class="plotlyFrame" src="./emb_dist_matrix.html" style="border: none;">
</iframe>
</div>
<div id="tabset-2-3">
<div style="margin-top: 5px;">

</div>
<div style="font-size: 0.6em;">
<ul>
<li>The absolute difference matrix is the absolute difference of the sequence distance matrix and the embedding distance matrix.</li>
</ul>
</div>
<iframe class="plotlyFrame" src="./abs_diff_dist_matrix.html" style="border: none;">
</iframe>
</div>
<div id="tabset-2-4">
<div style="margin-top: 18px;">

</div>
<div style="font-size: 0.6em;">
<ul>
<li>The difference distance matrix is the sequence distance matrix minus the embedding distance matrix.</li>
</ul>
</div>
<div style="margin-top: 18px;">

</div>
<iframe class="plotlyFrame" src="./diff_dist_matrix.html" style="border: none;">
</iframe>
</div>
</div>
</div>
</section>
<section id="radial-dendrograms-1" class="slide level4">
<h4>Radial Dendrograms (1)</h4>
<div>
<div style="font-size: 0.6em;">
<ul>
<li class="fragment">Radial dendrograms show who easily sequences cluster by their variant.</li>
<li class="fragment">Figures 12 and 13 in the next two slides show:
<ul>
<li class="fragment">Radial dendrograms using sequence data does not perfectly clusters sequences by variants.</li>
<li class="fragment">Radial dendrograms using the embedded data perfectly clusters sequences by variants.</li>
</ul></li>
</ul>
</div>
</div>
</section>
<section id="radial-dendrograms-2" class="slide level4">
<h4>Radial Dendrograms (2)</h4>

<img data-src="images/Radial_dendrograms.png" class="r-stretch"></section>

<section id="separability" class="title-slide slide level3 center">
<h3><span style="color:#A9D0E3;">Separability</span></h3>

</section>

<section id="principal-component-analysis" class="title-slide slide level3 center">
<h3>Principal Component Analysis</h3>
<div style="margin-top: 70px;">

</div>
<iframe class="plotlyFrame" src="./v_pc_plotly.html" style="border: none;">
</iframe>
</section>


<section id="extrinsic-evaluation-results" class="title-slide slide level2 center">
<h2><span style="color:#6FB1D1;">Extrinsic Evaluation Results</span></h2>

</section>
<section id="variant-classification" class="title-slide slide level3 center">
<h3><span style="color:#A9D0E3;">Variant Classification</span></h3>

</section>
<section id="variant-embedding-classification-1" class="slide level4">
<h4><span style="font-size: 0.8em;">Variant Embedding Classification (1)</span></h4>
<div>
<div style="font-size: 0.5em;">
<ul>
<li class="fragment">Variant classification with a CART learner had a 97.71% on aligned sequence embeddings.</li>
</ul>
</div>
</div>
<div class="fragment">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="images/Variant_CART_tree.png" style="width:67.0%"></p>
<figcaption>Figure 17: Variant Missclassifcations</figcaption>
</figure>
</div>
</div>
</section>
<section id="variant-embedding-classification-2" class="slide level4">
<h4>Variant Embedding Classification (2)</h4>
<div style="text-align:center; font-size: 0.5em;">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="images/Variant_missclassifcations.png" style="width:40.0%"></p>
<figcaption>Figure 17: Variant Missclassifcations</figcaption>
</figure>
</div>
</div>
</section>
<section id="variant-one-hot-encoding-classification" class="slide level4">
<h4>Variant One-Hot-Encoding Classification</h4>
<div class="fragment">
<div>
<div style="font-size: 0.6em;">
<ul>
<li class="fragment">Variant classification with a CART learner had a 10.28% on One-Hot-Encoded sequences.</li>
</ul>
</div>
</div>
</div>
<div class="fragment">
<p><object data="images/CART_variant_binary_cropped.pdf" type="application/pdf" width="900" height="400"><iframe src="images/CART_variant_binary_cropped.pdf" width="900" height="400">This browser does not support PDFs. Please download the PDF to view it: <a href="images/CART_variant_binary_cropped.pdf">Download PDF</a></iframe></object></p>
<aside class="notes">
<ul>
<li><p>one-hot encoding Covid sequences (assuming all sequences are of length 30,000) creates a one dimensional array of <span class="math inline">\(n \times 30,000 \times 4\)</span>, 4 columns representing 4 different bases</p></li>
<li><p>A: ([1, 0, 0, 0])</p></li>
<li><p>C: ([0, 1, 0, 0])</p></li>
<li><p>G: ([0, 0, 1, 0])</p></li>
<li><p>T: ([0, 0, 0, 1])</p></li>
</ul>
<p>where a there is 30,000 rows to represent all 30,000 nucleotides in a Covid sequence; thus with n matrices of <span class="math inline">\(30,000 \times 4\)</span> comprising the multidimensional array for each patient (ie each whole genome sequence)</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
<!--
### Protein Classification

. . .

::: panel-tabset
#### Embedding


::: {style="font-size: 0.5em;"}
- 100% classification rate using the all 512 dimensions from the GenSLM embedding.
:::










#### PCA Reduced Embedding

::: {style="font-size: 0.5em;"}
- 100% classification rate using the PCA reduced feature matrix with 28 dimensions.
:::









:::
-->
</div>
</section>


<section id="thesis-question-1" class="title-slide slide level2 center">
<h2>Thesis Question</h2>
<div style="text-align:center; font-size: 0.85em;">
<p>How easily can information be extracted from GenSLM embeddings while maintaining its original integrity; and what is the quality of the produced vector embeddings?</p>
</div>
</section>

<section id="conclusion" class="title-slide slide level2 center">
<h2>Conclusion</h2>
<div>
<div style="font-size: 0.6em;">
<ul>
<li class="fragment">The study assessed GenSLM’s embeddings for quality using intrinsic and extrinsic evaluations, focusing on redundancy, separability, and information preservation.</li>
<li class="fragment">GenSLM demonstrated strong performance in classification tasks, with a 97.71% success rate for variant classification and 100% for proteins using the CART learner, significantly outperforming One-Hot-Encoded matrices.</li>
<li class="fragment">While GenSLM excelled in separability, it showed less effectiveness in preserving fine-scale genetic differences and distorted the genetic distances between certain viral variants.</li>
<li class="fragment">The embeddings displayed considerable dimensional redundancy and only moderate success in preserving semantic distances, indicating areas for improvement.</li>
<li class="fragment">Future research should compare GenSLM with other neural network embedding algorithms and explore its utility in more diverse genomic analysis tasks to establish broader applicability.</li>
</ul>
</div>
</div>
</section>
</section>
<section id="thank-you" class="title-slide slide level1 center" data-background-color="#8FA9C2">
<h1><span class="center" style="color:white;">Thank you!!</span></h1>

</section>

<section id="appendix" class="title-slide slide level1 center" data-background-color="#8FA9C2">
<h1><span style="color:white;">Appendix</span></h1>

</section>

<section id="section-1" class="title-slide slide level1 center">
<h1></h1>
<p><object data="DHintz_PlanB_Submission.pdf" type="application/pdf" width="1000" height="650"><iframe src="DHintz_PlanB_Submission.pdf" width="1000" height="650">This browser does not support PDFs. Please download the PDF to view it: <a href="DHintz_PlanB_Submission.pdf">Download PDF</a></iframe></object></p>
</section>

<section>
<section id="acknowledgements" class="title-slide slide level1 center" data-background-color="#8FA9C2">
<h1><svg aria-hidden="true" role="img" viewbox="0 0 576 512" style="height:1em;width:1.12em;vertical-align:-0.125em;margin-left:auto;margin-right:auto;font-size:inherit;fill:white;overflow:visible;position:relative;"><path d="M163.9 136.9c-29.4-29.8-29.4-78.2 0-108s77-29.8 106.4 0l17.7 18 17.7-18c29.4-29.8 77-29.8 106.4 0s29.4 78.2 0 108L310.5 240.1c-6.2 6.3-14.3 9.4-22.5 9.4s-16.3-3.1-22.5-9.4L163.9 136.9zM568.2 336.3c13.1 17.8 9.3 42.8-8.5 55.9L433.1 485.5c-23.4 17.2-51.6 26.5-80.7 26.5H192 32c-17.7 0-32-14.3-32-32V416c0-17.7 14.3-32 32-32H68.8l44.9-36c22.7-18.2 50.9-28 80-28H272h16 64c17.7 0 32 14.3 32 32s-14.3 32-32 32H288 272c-8.8 0-16 7.2-16 16s7.2 16 16 16H392.6l119.7-88.2c17.8-13.1 42.8-9.3 55.9 8.5zM193.6 384l0 0-.9 0c.3 0 .6 0 .9 0z"></path></svg> <span style="color:white; font-size: 70px;">Acknowledgements</span></h1>
<div>
<ul>
<li class="fragment">Committee members</li>
<li class="fragment">Cohort</li>
</ul>
</div>
</section>
<section id="word-embeddings" class="title-slide slide level2 center">
<h2>Word Embeddings</h2>
<div>
<div style="font-size: 0.6em;">
<ul>
<li class="fragment">Before we introduce GenSLM, lets look at a simpler application in natural language.
<ul>
<li class="fragment">If we consider English text, <strong>How can we measure the similarity between words?</strong></li>
<li class="fragment">For example, what is the semantic difference between the words “King” and “Woman”.</li>
<li class="fragment">Let’s demonstrate the semantic relationships of “King”, “Queen”, “Man”, “Woman”.</li>
</ul></li>
</ul>
</div>
</div>
<aside class="notes">
<ul>
<li>Embedding algorithms need data (a corpus of text) to construct numeric vectors relating words and their meaning.</li>
</ul>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>

<section id="word2vec-example" class="title-slide slide level2 center">
<h2>Word2vec Example</h2>
<div class="fragment">
<div id="qwebr-insertion-location-2"></div>
<noscript>Please enable JavaScript to experience the dynamic code cell content on this page.</noscript>
<div class="cell small-code">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode numberSource r small-code number-lines code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href=""></a><span class="do">## Pre-Ran Code ##</span></span>
<span id="cb1-2"><a href=""></a><span class="fu">library</span>(word2vec)</span>
<span id="cb1-3"><a href=""></a><span class="co"># using British National Corpus http://vectors.nlpl.eu/repository/</span></span>
<span id="cb1-4"><a href=""></a>model <span class="ot">&lt;-</span> <span class="fu">read.word2vec</span>(<span class="st">"British_National_Corpus_Vector_size_300_skipgram"</span>, <span class="at">normalize =</span> <span class="cn">TRUE</span>)</span>
<span id="cb1-5"><a href=""></a>embedding <span class="ot">&lt;-</span> <span class="fu">predict</span>(model,<span class="at">newdata =</span> <span class="fu">c</span>(<span class="st">"king_NOUN"</span>,<span class="st">"man_NOUN"</span>,<span class="st">"woman_NOUN"</span>,<span class="st">"queen_NOUN"</span>),<span class="at">type=</span><span class="st">"embedding"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div style="margin-top: 20px;">

</div>
</div>
<div class="fragment">
<div id="qwebr-insertion-location-3"></div>
<noscript>Please enable JavaScript to experience the dynamic code cell content on this page.</noscript>
</div>
</section>

<section id="plotting-word-embeddings" class="title-slide slide level2 center">
<h2>Plotting Word Embeddings</h2>
<div class="fragment">
<div style="font-size: 0.6em;">
<ul>
<li>We can Plot the vectors shown in the previous slide to demonstrate how word2vec produces embeddings that measures semantic similarity between words.</li>
</ul>
</div>
</div>
<div class="fragment">
<div style="text-align:center; font-size: 0.5em;">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="images/King_Man_Woman_Queen.png" style="width:70.0%"></p>
<figcaption>Figure 4: Word Embedding Vectors</figcaption>
</figure>
</div>
</div>
</div>
</section>
<section id="kullbackleibler-divergence-kld" class="slide level4">
<h4>Kullback–Leibler Divergence (KLD)</h4>
<div style="text-align:center; font-size: 0.5em;">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="images/CKL_vs_DKL_ver2.png" style="width:85.0%"></p>
<figcaption>Figure 15: CKL vs DKL</figcaption>
</figure>
</div>
</div>
</section>
</section>
<section id="references" class="title-slide slide level1 smaller scrollable" data-background-color="#8FA9C2">
<h1><span class="center" style="color:#464D58;">References</span></h1>
<div class="quarto-auto-generated-content">
<p><img src="images/quarto.png" class="slide-logo"></p>
<div class="footer footer-default">

</div>
</div>
<div class="reveal-header">
<div class="header-logo" data-footer-logo-link="https://quarto.org">
<img data-src="">
</div>
<div class="sc-title">
<p> </p>
</div>
<div class="header-text">
<p> </p>
</div>
<div class="sb-title">
<p> </p>
</div>
</div>
<div id="refs" class="references csl-bib-body" role="list">
<div id="ref-NHGRI2024" class="csl-entry" role="listitem">
<div class="csl-left-margin">1. </div><div class="csl-right-inline">National Human Genome Research Institute (2024) <a href="https://www.genome.gov/genetics-glossary/Base-Pair">Base pair</a></div>
</div>
<div id="ref-kandwal2023genetic" class="csl-entry" role="listitem">
<div class="csl-left-margin">2. </div><div class="csl-right-inline">Kandwal S, Fayne D (2023) Genetic conservation across SARS-CoV-2 non-structural proteins–insights into possible targets for treatment of future viral outbreaks. Virology</div>
</div>
<div id="ref-genslms2023" class="csl-entry" role="listitem">
<div class="csl-left-margin">3. </div><div class="csl-right-inline">Zvyagin M, Brace A, Hippe K, et al (2023) GenSLMs: Genome-scale language models reveal SARS-CoV-2 evolutionary dynamics. The International Journal of High Performance Computing Applications 37(6):683–705</div>
</div>
<div id="ref-lavravc2021representation" class="csl-entry" role="listitem">
<div class="csl-left-margin">4. </div><div class="csl-right-inline">Lavrač N, Podpečan V, Robnik-Šikonja M (2021) Representation learning: Propositionalization and embeddings. Springer</div>
</div>
</div>
</section>
    </div>
  </div>

  <script>window.backupDefine = window.define; window.define = undefined;</script>
  <script src="index_files/libs/revealjs/dist/reveal.js"></script>
  <!-- reveal.js plugins -->
  <script src="index_files/libs/revealjs/plugin/quarto-line-highlight/line-highlight.js"></script>
  <script src="index_files/libs/revealjs/plugin/pdf-export/pdfexport.js"></script>
  <script src="index_files/libs/revealjs/plugin/reveal-menu/menu.js"></script>
  <script src="index_files/libs/revealjs/plugin/reveal-menu/quarto-menu.js"></script>
  <script src="index_files/libs/revealjs/plugin/reveal-chalkboard/plugin.js"></script>
  <script src="index_files/libs/revealjs/plugin/quarto-support/support.js"></script>
  

  <script src="index_files/libs/revealjs/plugin/notes/notes.js"></script>
  <script src="index_files/libs/revealjs/plugin/search/search.js"></script>
  <script src="index_files/libs/revealjs/plugin/zoom/zoom.js"></script>
  <script src="index_files/libs/revealjs/plugin/math/math.js"></script>
  <script>window.define = window.backupDefine; window.backupDefine = undefined;</script>

  <script>

      // Full list of configuration options available at:
      // https://revealjs.com/config/
      Reveal.initialize({
'controlsAuto': true,
'previewLinksAuto': false,
'pdfSeparateFragments': false,
'autoAnimateEasing': "ease",
'autoAnimateDuration': 1,
'autoAnimateUnmatched': true,
'menu': {"side":"left","useTextContentForMissingTitles":true,"markers":false,"loadIcons":false,"custom":[{"title":"Tools","icon":"<i class=\"fas fa-gear\"></i>","content":"<ul class=\"slide-menu-items\">\n<li class=\"slide-tool-item active\" data-item=\"0\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.fullscreen(event)\"><kbd>f</kbd> Fullscreen</a></li>\n<li class=\"slide-tool-item\" data-item=\"1\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.speakerMode(event)\"><kbd>s</kbd> Speaker View</a></li>\n<li class=\"slide-tool-item\" data-item=\"2\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.overview(event)\"><kbd>o</kbd> Slide Overview</a></li>\n<li class=\"slide-tool-item\" data-item=\"3\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.togglePdfExport(event)\"><kbd>e</kbd> PDF Export Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"4\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleChalkboard(event)\"><kbd>b</kbd> Toggle Chalkboard</a></li>\n<li class=\"slide-tool-item\" data-item=\"5\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleNotesCanvas(event)\"><kbd>c</kbd> Toggle Notes Canvas</a></li>\n<li class=\"slide-tool-item\" data-item=\"6\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.downloadDrawings(event)\"><kbd>d</kbd> Download Drawings</a></li>\n<li class=\"slide-tool-item\" data-item=\"7\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.keyboardHelp(event)\"><kbd>?</kbd> Keyboard Help</a></li>\n</ul>"}],"openButton":true},
'chalkboard': {"buttons":true},
'smaller': false,
 
        // Display controls in the bottom right corner
        controls: false,

        // Help the user learn the controls by providing hints, for example by
        // bouncing the down arrow when they first encounter a vertical slide
        controlsTutorial: false,

        // Determines where controls appear, "edges" or "bottom-right"
        controlsLayout: 'edges',

        // Visibility rule for backwards navigation arrows; "faded", "hidden"
        // or "visible"
        controlsBackArrows: 'faded',

        // Display a presentation progress bar
        progress: true,

        // Display the page number of the current slide
        slideNumber: 'c/t',

        // 'all', 'print', or 'speaker'
        showSlideNumber: 'all',

        // Add the current slide number to the URL hash so that reloading the
        // page/copying the URL will return you to the same slide
        hash: true,

        // Start with 1 for the hash rather than 0
        hashOneBasedIndex: false,

        // Flags if we should monitor the hash and change slides accordingly
        respondToHashChanges: true,

        // Push each slide change to the browser history
        history: true,

        // Enable keyboard shortcuts for navigation
        keyboard: true,

        // Enable the slide overview mode
        overview: true,

        // Disables the default reveal.js slide layout (scaling and centering)
        // so that you can use custom CSS layout
        disableLayout: false,

        // Vertical centering of slides
        center: false,

        // Enables touch navigation on devices with touch input
        touch: true,

        // Loop the presentation
        loop: false,

        // Change the presentation direction to be RTL
        rtl: false,

        // see https://revealjs.com/vertical-slides/#navigation-mode
        navigationMode: 'linear',

        // Randomizes the order of slides each time the presentation loads
        shuffle: false,

        // Turns fragments on and off globally
        fragments: true,

        // Flags whether to include the current fragment in the URL,
        // so that reloading brings you to the same fragment position
        fragmentInURL: false,

        // Flags if the presentation is running in an embedded mode,
        // i.e. contained within a limited portion of the screen
        embedded: false,

        // Flags if we should show a help overlay when the questionmark
        // key is pressed
        help: true,

        // Flags if it should be possible to pause the presentation (blackout)
        pause: true,

        // Flags if speaker notes should be visible to all viewers
        showNotes: false,

        // Global override for autoplaying embedded media (null/true/false)
        autoPlayMedia: null,

        // Global override for preloading lazy-loaded iframes (null/true/false)
        preloadIframes: null,

        // Number of milliseconds between automatically proceeding to the
        // next slide, disabled when set to 0, this value can be overwritten
        // by using a data-autoslide attribute on your slides
        autoSlide: 0,

        // Stop auto-sliding after user input
        autoSlideStoppable: true,

        // Use this method for navigation when auto-sliding
        autoSlideMethod: null,

        // Specify the average time in seconds that you think you will spend
        // presenting each slide. This is used to show a pacing timer in the
        // speaker view
        defaultTiming: null,

        // Enable slide navigation via mouse wheel
        mouseWheel: false,

        // The display mode that will be used to show slides
        display: 'block',

        // Hide cursor if inactive
        hideInactiveCursor: true,

        // Time before the cursor is hidden (in ms)
        hideCursorTime: 5000,

        // Opens links in an iframe preview overlay
        previewLinks: false,

        // Transition style (none/fade/slide/convex/concave/zoom)
        transition: 'slide',

        // Transition speed (default/fast/slow)
        transitionSpeed: 'default',

        // Transition style for full page slide backgrounds
        // (none/fade/slide/convex/concave/zoom)
        backgroundTransition: 'none',

        // Number of slides away from the current that are visible
        viewDistance: 3,

        // Number of slides away from the current that are visible on mobile
        // devices. It is advisable to set this to a lower number than
        // viewDistance in order to save resources.
        mobileViewDistance: 2,

        // The "normal" size of the presentation, aspect ratio will be preserved
        // when the presentation is scaled to fit different resolutions. Can be
        // specified using percentage units.
        width: 1050,

        height: 700,

        // Factor of the display size that should remain empty around the content
        margin: 0.1,

        math: {
          mathjax: 'https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js',
          config: 'TeX-AMS_HTML-full',
          tex2jax: {
            inlineMath: [['\\(','\\)']],
            displayMath: [['\\[','\\]']],
            balanceBraces: true,
            processEscapes: false,
            processRefs: true,
            processEnvironments: true,
            preview: 'TeX',
            skipTags: ['script','noscript','style','textarea','pre','code'],
            ignoreClass: 'tex2jax_ignore',
            processClass: 'tex2jax_process'
          },
        },

        // reveal.js plugins
        plugins: [QuartoLineHighlight, PdfExport, RevealMenu, RevealChalkboard, QuartoSupport,

          RevealMath,
          RevealNotes,
          RevealSearch,
          RevealZoom
        ]
      });
    </script>
    
    <script>
      // htmlwidgets need to know to resize themselves when slides are shown/hidden.
      // Fire the "slideenter" event (handled by htmlwidgets.js) when the current
      // slide changes (different for each slide format).
      (function () {
        // dispatch for htmlwidgets
        function fireSlideEnter() {
          const event = window.document.createEvent("Event");
          event.initEvent("slideenter", true, true);
          window.document.dispatchEvent(event);
        }

        function fireSlideChanged(previousSlide, currentSlide) {
          fireSlideEnter();

          // dispatch for shiny
          if (window.jQuery) {
            if (previousSlide) {
              window.jQuery(previousSlide).trigger("hidden");
            }
            if (currentSlide) {
              window.jQuery(currentSlide).trigger("shown");
            }
          }
        }

        // hookup for slidy
        if (window.w3c_slidy) {
          window.w3c_slidy.add_observer(function (slide_num) {
            // slide_num starts at position 1
            fireSlideChanged(null, w3c_slidy.slides[slide_num - 1]);
          });
        }

      })();
    </script>

    <script id="quarto-html-after-body" type="application/javascript">
    window.document.addEventListener("DOMContentLoaded", function (event) {
      const toggleBodyColorMode = (bsSheetEl) => {
        const mode = bsSheetEl.getAttribute("data-mode");
        const bodyEl = window.document.querySelector("body");
        if (mode === "dark") {
          bodyEl.classList.add("quarto-dark");
          bodyEl.classList.remove("quarto-light");
        } else {
          bodyEl.classList.add("quarto-light");
          bodyEl.classList.remove("quarto-dark");
        }
      }
      const toggleBodyColorPrimary = () => {
        const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
        if (bsSheetEl) {
          toggleBodyColorMode(bsSheetEl);
        }
      }
      toggleBodyColorPrimary();  
      const tabsets =  window.document.querySelectorAll(".panel-tabset-tabby")
      tabsets.forEach(function(tabset) {
        const tabby = new Tabby('#' + tabset.id);
      });
      const isCodeAnnotation = (el) => {
        for (const clz of el.classList) {
          if (clz.startsWith('code-annotation-')) {                     
            return true;
          }
        }
        return false;
      }
      const clipboard = new window.ClipboardJS('.code-copy-button', {
        text: function(trigger) {
          const codeEl = trigger.previousElementSibling.cloneNode(true);
          for (const childEl of codeEl.children) {
            if (isCodeAnnotation(childEl)) {
              childEl.remove();
            }
          }
          return codeEl.innerText;
        }
      });
      clipboard.on('success', function(e) {
        // button target
        const button = e.trigger;
        // don't keep focus
        button.blur();
        // flash "checked"
        button.classList.add('code-copy-button-checked');
        var currentTitle = button.getAttribute("title");
        button.setAttribute("title", "Copied!");
        let tooltip;
        if (window.bootstrap) {
          button.setAttribute("data-bs-toggle", "tooltip");
          button.setAttribute("data-bs-placement", "left");
          button.setAttribute("data-bs-title", "Copied!");
          tooltip = new bootstrap.Tooltip(button, 
            { trigger: "manual", 
              customClass: "code-copy-button-tooltip",
              offset: [0, -8]});
          tooltip.show();    
        }
        setTimeout(function() {
          if (tooltip) {
            tooltip.hide();
            button.removeAttribute("data-bs-title");
            button.removeAttribute("data-bs-toggle");
            button.removeAttribute("data-bs-placement");
          }
          button.setAttribute("title", currentTitle);
          button.classList.remove('code-copy-button-checked');
        }, 1000);
        // clear code selection
        e.clearSelection();
      });
      function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
        const config = {
          allowHTML: true,
          maxWidth: 500,
          delay: 100,
          arrow: false,
          appendTo: function(el) {
              return el.closest('section.slide') || el.parentElement;
          },
          interactive: true,
          interactiveBorder: 10,
          theme: 'light-border',
          placement: 'bottom-start',
        };
        if (contentFn) {
          config.content = contentFn;
        }
        if (onTriggerFn) {
          config.onTrigger = onTriggerFn;
        }
        if (onUntriggerFn) {
          config.onUntrigger = onUntriggerFn;
        }
          config['offset'] = [0,0];
          config['maxWidth'] = 700;
        window.tippy(el, config); 
      }
      const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
      for (var i=0; i<noterefs.length; i++) {
        const ref = noterefs[i];
        tippyHover(ref, function() {
          // use id or data attribute instead here
          let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
          try { href = new URL(href).hash; } catch {}
          const id = href.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          return note.innerHTML;
        });
      }
      const findCites = (el) => {
        const parentEl = el.parentElement;
        if (parentEl) {
          const cites = parentEl.dataset.cites;
          if (cites) {
            return {
              el,
              cites: cites.split(' ')
            };
          } else {
            return findCites(el.parentElement)
          }
        } else {
          return undefined;
        }
      };
      var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
      for (var i=0; i<bibliorefs.length; i++) {
        const ref = bibliorefs[i];
        const citeInfo = findCites(ref);
        if (citeInfo) {
          tippyHover(citeInfo.el, function() {
            var popup = window.document.createElement('div');
            citeInfo.cites.forEach(function(cite) {
              var citeDiv = window.document.createElement('div');
              citeDiv.classList.add('hanging-indent');
              citeDiv.classList.add('csl-entry');
              var biblioDiv = window.document.getElementById('ref-' + cite);
              if (biblioDiv) {
                citeDiv.innerHTML = biblioDiv.innerHTML;
              }
              popup.appendChild(citeDiv);
            });
            return popup.innerHTML;
          });
        }
      }
    });
    </script>
    <script type="module">
    // Handle cell initialization initialization
    qwebrCellDetails.map(
      (entry) => {
        // Handle the creation of the element
        qwebrCreateHTMLElement(entry);
        // In the event of interactive, initialize the monaco editor
        if (entry.options.context == EvalTypes.Interactive) {
          qwebrCreateMonacoEditorInstance(entry);
        }
      }
    );

    // Identify non-interactive cells (in order)
    const filteredEntries = qwebrCellDetails.filter(entry => {
      const contextOption = entry.options && entry.options.context;
      return ['output', 'setup'].includes(contextOption) || (contextOption == "interactive" && entry.options && entry.options.autorun === 'true');
    });

    // Condition non-interactive cells to only be run after webR finishes its initialization.
    qwebrInstance.then(
      async () => {
        const nHiddenCells = filteredEntries.length;
        var currentHiddenCell = 0;


        // Modify button state
        qwebrSetInteractiveButtonState(`🟡 Running hidden code cells ...`, false);

        // Begin processing non-interactive sections
        // Due to the iteration policy, we must use a for() loop.
        // Otherwise, we would need to switch to using reduce with an empty
        // starting promise
        for (const entry of filteredEntries) {

          // Determine cell being examined
          currentHiddenCell = currentHiddenCell + 1;
          const formattedMessage = `Evaluating hidden cell ${currentHiddenCell} out of ${nHiddenCells}`;

          // Update the document status header
          if (qwebrShowStartupMessage) {
            qwebrUpdateStatusHeader(formattedMessage);
          }

          // Display the update in non-active areas
          qwebrUpdateStatusMessage(formattedMessage);

          // Extract details on the active cell
          const evalType = entry.options.context;
          const cellCode = entry.code;
          const qwebrCounter = entry.id;

          if (['output', 'setup'].includes(evalType)) {
            // Disable further global status updates
            const activeContainer = document.getElementById(`qwebr-non-interactive-loading-container-${qwebrCounter}`);
            activeContainer.classList.remove('qwebr-cell-needs-evaluation');
            activeContainer.classList.add('qwebr-cell-evaluated');

            // Update status on the code cell
            const activeStatus = document.getElementById(`qwebr-status-text-${qwebrCounter}`);
            activeStatus.innerText = " Evaluating hidden code cell...";
            activeStatus.classList.remove('qwebr-cell-needs-evaluation');
            activeStatus.classList.add('qwebr-cell-evaluated');
          }

          switch (evalType) {
            case 'interactive':
              // TODO: Make this more standardized.
              // At the moment, we're overriding the interactive status update by pretending its
              // output-like. 
              const tempOptions = entry.options;
              tempOptions["context"] = "output"
              // Run the code in a non-interactive state that is geared to displaying output
              await qwebrExecuteCode(`${cellCode}`, qwebrCounter, tempOptions);
              break;
            case 'output':
              // Run the code in a non-interactive state that is geared to displaying output
              await qwebrExecuteCode(`${cellCode}`, qwebrCounter, entry.options);
              break;
            case 'setup':
              const activeDiv = document.getElementById(`qwebr-noninteractive-setup-area-${qwebrCounter}`);
              // Run the code in a non-interactive state with all output thrown away
              await mainWebR.evalRVoid(`${cellCode}`);
              break;
            default: 
              break; 
          }

          if (['output', 'setup'].includes(evalType)) {
            // Disable further global status updates
            const activeContainer = document.getElementById(`qwebr-non-interactive-loading-container-${qwebrCounter}`);
            // Disable visibility
            activeContainer.style.visibility = 'hidden';
            activeContainer.style.display = 'none';
          }
        }
      }
    ).then(
      () => {
        // Release document status as ready

        if (qwebrShowStartupMessage) {
          qwebrStartupMessage.innerText = "🟢 Ready!"
        }
      
        qwebrSetInteractiveButtonState(
          `<i class="fa-solid fa-play qwebr-icon-run-code"></i> <span>Run Code</span>`, 
          true
        );  
      }
    );
    </script>
    

</body></html>